# NHANES Diabetes Risk Data Exploration

In this notebook, I explore NHANES data from the following survey cycles:
- 2011–2012 
- 2013–2014 
- 2015–2016 
- 2017–2018 
It contains 32 NHANES .xpt source files across the survey cycle with the combined raw file size of ~ 65.20 MB  
Each component contains a different type of participant information

`DEMO`: Demographics - Age, sex, race/ethnicity, education, income, and survey design variables 

`GHB` : Glycohemoglobin - Laboratory-measured HbA1c; source of the machine-learning target 

`BMX` : Body Measures - BMI, weight, height, waist circumference, and related measurements 

`DIQ` : Diabetes Questionnaire - Self-reported diabetes diagnosis and related diabetes information 

`PAQ` : Physical Activity - Physical activity, exercise, and sedentary behavior 

`SMQ` : Smoking - Cigarette Use - Lifetime and current cigarette-smoking behavior 

`SLQ` : Sleep Disorders - Sleep duration and selected sleep-related information 

`BPQ` : Blood Pressure & Cholesterol Questionnaire - Self-reported hypertension and cholesterol-related information 

These files are in SAS transport format where each survey cycle uses suffix in the filename to represent the year.

G: 2011-2012, H: 2013–2014, I: 2015–2016, J: 2017–2018


  

| Component                            | Variables we care about                                                                               | 2011–12                                                                                                          | 2013–14                                                                                                          | 2015–16                                                                                                          | 2017–18                                                                                                          |
| ------------------------------------ | ----------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- |
| **DEMO : Demographics**              | `SEQN`, `RIDAGEYR`, `RIAGENDR`, `RIDRETH3`, `DMDEDUC2`, `INDFMPIR`, `WTMEC2YR`, `SDMVPSU`, `SDMVSTRA` | [DEMO_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/DEMO_G.htm) | [DEMO_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/DEMO_H.htm) | [DEMO_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/DEMO_I.htm) | [DEMO_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/DEMO_J.htm) |
| **GHB : HbA1c**                      | `LBXGH`                                                                                               | [GHB_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/GHB_G.htm)   | [GHB_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/GHB_H.htm)   | [GHB_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/GHB_I.htm)   | [GHB_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/GHB_J.htm)   |
| **BMX : Body Measures**              | `BMXBMI`, `BMXWAIST`, `BMXWT`                                                                         | [BMX_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/BMX_G.htm)   | [BMX_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/BMX_H.htm)   | [BMX_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/BMX_I.htm )   | [BMX_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/BMX_J.htm )   |
| **DIQ : Diabetes**                   | `DIQ010`                                                                                              | [DIQ_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/DIQ_G.htm )   | [DIQ_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/DIQ_H.htm )   | [DIQ_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/DIQ_I.htm )   | [DIQ_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/DIQ_J.htm )   |
| **PAQ : Physical Activity**          | `PAQ650`, `PAQ655`, `PAD660`, `PAQ665`, `PAQ670`, `PAD675`, `PAD680`                                  | [PAQ_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/PAQ_G.htm )   | [PAQ_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/PAQ_H.htm )   | [PAQ_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/PAQ_I.htm)   | [PAQ_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/PAQ_J.htm )   |
| **SMQ : Smoking**                    | `SMQ020`, `SMQ040`                                                                                    | [SMQ_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/SMQ_G.htm )   | [SMQ_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/SMQ_H.htm )   | [SMQ_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/SMQ_I.htm )   | [SMQ_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/SMQ_J.htm)   |
| **SLQ : Sleep**                      | `SLD010H` → older cycles; `SLD012` → newer cycles                                                     | [SLQ_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/SLQ_G.htm)   | [SLQ_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/SLQ_H.htm)   | [SLQ_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/SLQ_I.htm)   | [SLQ_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/SLQ_J.htm)   |
| **BPQ : Blood Pressure/Cholesterol** | `BPQ020`, `BPQ080`                                                                                    | [BPQ_G codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2011/DataFiles/BPQ_G.htm)   | [BPQ_H codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2013/DataFiles/BPQ_H.htm)   | [BPQ_I codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2015/DataFiles/BPQ_I.htm)   | [BPQ_J codebook](https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/BPQ_J.htm)   |


## 1. Libraries and the Data

In [2]:
from pathlib import Path
import shutil
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import requests
import sklearn

print("pandas:", pd.__version__)
print("numpy:", np.__version__)
print("scikit-learn:", sklearn.__version__)


pandas: 2.3.3
numpy: 2.3.4
scikit-learn: 1.9.1


# 2. Importing the Data

In [6]:
# Setting up the path 
data_dir = Path("../data")

#  NHANES components to be used
components = [
    "DEMO",  # Demographics
    "GHB",   # Glycohemoglobin / HbA1c
    "BMX",   # Body Measures
    "DIQ",   # Diabetes Questionnaire
    "PAQ",   # Physical Activity
    "SMQ",   # Smoking - Cigarette Use
    "SLQ",   # Sleep Disorders
    "BPQ"    # Blood Pressure & Cholesterol Questionnaire
]

# NHANES survey cycles
cycles = {
    "G": {
        "year": "2011",
        "cycle": "2011-2012"
    },
    "H": {
        "year": "2013",
        "cycle": "2013-2014"
    },
    "I": {
        "year": "2015",
        "cycle": "2015-2016"
    },
    "J": {
        "year": "2017",
        "cycle": "2017-2018"
    }
}

#  Downloading all required NHANES files
for suffix, cycle_info in cycles.items():

    year = cycle_info["year"]
    cycle_name = cycle_info["cycle"]

    print("\n" + "=" * 60)
    print(f"NHANES {cycle_name}")
    print("=" * 60)

    # going through each data component.
    for component in components:

        # Building the expected filename i.e DEMO_G.xpt
        filename = f"{component}_{suffix}.xpt"

        # local path where the file should be saved.
        output_path = data_dir / filename

        # Check if the file already exists
   

        if output_path.exists():

            # file size for reference.
            size_mb = output_path.stat().st_size / (1024 ** 2)

            print(
                f"EXISTS      {filename:<12} "
                f"{size_mb:.3f} MB"
            )

            # Skipping  if the file is already present.
            continue


        # official CDC download URL
        url = (
            f"https://wwwn.cdc.gov/Nchs/Data/Nhanes/"
            f"Public/{year}/DataFiles/{filename}"
        )

        #print(f"Downloading {filename} ...")

        try:

            # Requesting the files from the CDC website.
            response = requests.get(
                url,
                timeout=60
            )

            # Raising an error if the server returned
            response.raise_for_status()


            # checking CDC returns an XPT file and not an HTML error page.
            content = response.content

            first_bytes = content.lstrip().lower()[:100]

            if (
                first_bytes.startswith(b"<!doctype")
                or first_bytes.startswith(b"<html")
            ):
                print(
                    f"ERROR       {filename:<12} "
                    f"CDC returned a webpage instead of XPT data."
                )
                continue

            # Saving the downloaded file
            output_path.write_bytes(content)

            # Calculating file size 
            size_mb = output_path.stat().st_size / (1024 ** 2)

            print(
                f"DOWNLOADED  {filename:<12} "
                f"{size_mb:.3f} MB"
            )

        except requests.RequestException as error:

            print(
                f"FAILED      {filename:<12} "
                f"{error}"
            )

# Create list of all expected NHANES filenames
expected_files = [
    f"{component}_{suffix}.xpt"
    for suffix in cycles
    for component in components
]
        
#total file size           
active_files = [
    data_dir / filename
    for filename in expected_files
    if (data_dir / filename).exists()
]

total_size_mb = sum(
    file.stat().st_size
    for file in active_files
) / (1024 ** 2)

print(
    f"\nTotal size of active NHANES files: "
    f"{total_size_mb:.3f} MB"
)


NHANES 2011-2012
EXISTS      DEMO_G.xpt   3.580 MB
EXISTS      GHB_G.xpt    0.101 MB
EXISTS      BMX_G.xpt    1.857 MB
EXISTS      DIQ_G.xpt    3.794 MB
EXISTS      PAQ_G.xpt    1.463 MB
EXISTS      SMQ_G.xpt    1.857 MB
EXISTS      SLQ_G.xpt    0.190 MB
EXISTS      BPQ_G.xpt    0.709 MB

NHANES 2013-2014
EXISTS      DEMO_H.xpt   3.656 MB
EXISTS      GHB_H.xpt    0.107 MB
EXISTS      BMX_H.xpt    1.951 MB
EXISTS      DIQ_H.xpt    4.033 MB
EXISTS      PAQ_H.xpt    6.960 MB
EXISTS      SMQ_H.xpt    2.069 MB
EXISTS      SLQ_H.xpt    0.199 MB
EXISTS      BPQ_H.xpt    0.693 MB

NHANES 2015-2016
EXISTS      DEMO_I.xpt   3.582 MB
EXISTS      GHB_I.xpt    0.104 MB
EXISTS      BMX_I.xpt    1.897 MB
EXISTS      DIQ_I.xpt    3.953 MB
EXISTS      PAQ_I.xpt    6.651 MB
EXISTS      SMQ_I.xpt    2.557 MB
EXISTS      SLQ_I.xpt    0.352 MB
EXISTS      BPQ_I.xpt    0.533 MB

NHANES 2017-2018
EXISTS      DEMO_J.xpt   3.255 MB
EXISTS      GHB_J.xpt    0.099 MB
EXISTS      BMX_J.xpt    1.398 MB
EXISTS    

# 3. Rows/columns summary and compatibility

- summary table
- loading same component across the four cycle
- comparing the coloumn names across: variables present, missing or named differently and check for compatibility
- check for candidate variable
- refer to CDC codebooks
- find out the cross cycle features
- update variable names to make it meaningful
- combine 4 cycles after cleaning
- next step eda

### A. Demographic Data Compatibility Across Survey Cycles

Checking: the number of rows and columns in each cycle, variables that are present in all four cycles, variables that differ between cycles, and if the candidate demographic predictors are compatible across 2011–2018

We find from below that a total of 42 demographic variables were present in all four cycles. There are some additional household-reference-person variables changed between survey cycles but I don't plan on using them for machine learning so I can go ahead and combine the datasets. 








In [8]:
# Loading the demographics dataset

demo_g = pd.read_sas(
    data_dir / "DEMO_G.xpt",
    format="xport"
)

demo_h = pd.read_sas(
    data_dir / "DEMO_H.xpt",
    format="xport"
)

demo_i = pd.read_sas(
    data_dir / "DEMO_I.xpt",
    format="xport"
)

demo_j = pd.read_sas(
    data_dir / "DEMO_J.xpt",
    format="xport"
)
print("2011-2012:", demo_g.shape)
print("2013-2014:", demo_h.shape)
print("2015-2016:", demo_i.shape)
print("2017-2018:", demo_j.shape)


2011-2012: (9756, 48)
2013-2014: (10175, 47)
2015-2016: (9971, 47)
2017-2018: (9254, 46)


In [10]:
print("2011-2012 columns:")
print(demo_g.columns.tolist())

print("\n2013-2014 columns:")
print(demo_h.columns.tolist())

print("\n2015-2016 columns:")
print(demo_i.columns.tolist())

print("\n2017-2018 columns:")
print(demo_j.columns.tolist())

2011-2012 columns:
['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN', 'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'RIDEXAGY', 'RIDEXAGM', 'DMQMILIZ', 'DMQADFC', 'DMDBORN4', 'DMDCITZN', 'DMDYRSUS', 'DMDEDUC3', 'DMDEDUC2', 'DMDMARTL', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG', 'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA', 'WTINT2YR', 'WTMEC2YR', 'SDMVPSU', 'SDMVSTRA', 'INDHHIN2', 'INDFMIN2', 'INDFMPIR', 'DMDHHSIZ', 'DMDFMSIZ', 'DMDHHSZA', 'DMDHHSZB', 'DMDHHSZE', 'DMDHRGND', 'DMDHRAGE', 'DMDHRBR4', 'DMDHREDU', 'DMDHRMAR', 'DMDHSEDU']

2013-2014 columns:
['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN', 'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'RIDEXAGM', 'DMQMILIZ', 'DMQADFC', 'DMDBORN4', 'DMDCITZN', 'DMDYRSUS', 'DMDEDUC3', 'DMDEDUC2', 'DMDMARTL', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG', 'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA', 'DMDHHSIZ', 'DMDFMSIZ', 'DMDHHSZA', 'DMDHHSZB', 'DMD

#### Variables that exists in all four cycle

In [12]:
common_demo_columns = (
    set(demo_g.columns)
    & set(demo_h.columns)
    & set(demo_i.columns)
    & set(demo_j.columns)
)

print("Number of demographic variables common to all four cycles:")
print(len(common_demo_columns))

print("\nCommon demographic variables:")
print(sorted(common_demo_columns))

Number of demographic variables common to all four cycles:
42

Common demographic variables:
['AIALANGA', 'DMDBORN4', 'DMDCITZN', 'DMDEDUC2', 'DMDEDUC3', 'DMDFMSIZ', 'DMDHHSIZ', 'DMDHHSZA', 'DMDHHSZB', 'DMDHHSZE', 'DMDHRGND', 'DMDMARTL', 'DMDYRSUS', 'DMQADFC', 'DMQMILIZ', 'FIAINTRP', 'FIALANG', 'FIAPROXY', 'INDFMIN2', 'INDFMPIR', 'INDHHIN2', 'MIAINTRP', 'MIALANG', 'MIAPROXY', 'RIAGENDR', 'RIDAGEMN', 'RIDAGEYR', 'RIDEXAGM', 'RIDEXMON', 'RIDEXPRG', 'RIDRETH1', 'RIDRETH3', 'RIDSTATR', 'SDDSRVYR', 'SDMVPSU', 'SDMVSTRA', 'SEQN', 'SIAINTRP', 'SIALANG', 'SIAPROXY', 'WTINT2YR', 'WTMEC2YR']


#### Missing from each cycle

In [14]:
all_demo_columns = (
    set(demo_g.columns)
    | set(demo_h.columns)
    | set(demo_i.columns)
    | set(demo_j.columns)
)
missing_from_g = all_demo_columns - set(demo_g.columns)
missing_from_h = all_demo_columns - set(demo_h.columns)
missing_from_i = all_demo_columns - set(demo_i.columns)
missing_from_j = all_demo_columns - set(demo_j.columns)

print("Missing from 2011-2012:")
print(sorted(missing_from_g))

print("\nMissing from 2013-2014:")
print(sorted(missing_from_h))

print("\nMissing from 2015-2016:")
print(sorted(missing_from_i))

print("\nMissing from 2017-2018:")
print(sorted(missing_from_j))

Missing from 2011-2012:
['DMDHRAGZ', 'DMDHREDZ', 'DMDHRMAZ', 'DMDHSEDZ']

Missing from 2013-2014:
['DMDHRAGZ', 'DMDHREDZ', 'DMDHRMAZ', 'DMDHSEDZ', 'RIDEXAGY']

Missing from 2015-2016:
['DMDHRAGZ', 'DMDHREDZ', 'DMDHRMAZ', 'DMDHSEDZ', 'RIDEXAGY']

Missing from 2017-2018:
['DMDHRAGE', 'DMDHRBR4', 'DMDHREDU', 'DMDHRMAR', 'DMDHSEDU', 'RIDEXAGY']


In [16]:
# we want these demographics
candidate_demo_variables = [
    "SEQN",
    "RIDAGEYR",
    "RIAGENDR",
    "RIDRETH3",
    "DMDEDUC2",
    "INDFMPIR",
    "WTMEC2YR",
    "SDMVPSU",
    "SDMVSTRA"
]

demo_datasets = {
    "2011-2012": demo_g,
    "2013-2014": demo_h,
    "2015-2016": demo_i,
    "2017-2018": demo_j
}

demo_compatibility = []

for variable in candidate_demo_variables:
    
    row = {
        "variable": variable
    }
    
    for cycle, df in demo_datasets.items():
        row[cycle] = variable in df.columns
    
    demo_compatibility.append(row)

demo_compatibility_df = pd.DataFrame(demo_compatibility)

demo_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,RIDAGEYR,True,True,True,True
2,RIAGENDR,True,True,True,True
3,RIDRETH3,True,True,True,True
4,DMDEDUC2,True,True,True,True
5,INDFMPIR,True,True,True,True
6,WTMEC2YR,True,True,True,True
7,SDMVPSU,True,True,True,True
8,SDMVSTRA,True,True,True,True


###  B. Glycohemoglobin Data Compatibility

The Glycohemoglobin (GHB) component contains laboratory measurements of HbA1c, it will be used to construct the target variable for the machine learning analysis.

From below we know: Each dataset contains the same two variables: SEQN : unique participant identifier and LBXGH : laboratory-measured glycohemoglobin (HbA1c), percent as they are consistent it can be used for machine-learning target variable without requiring cross-cycle harmonization

There are 26,673 total Glycohemoglobin records and 25,159 participants with a valid HbA1c measurement. The 25,159 valid HbA1c measurements represent the maximum potential sample before additional filtering, including restricting the analysis to adults and excluding participants with a previously reported diabetes diagnosis.


In [18]:
# Loading glycohemoglobin datasets

ghb_g = pd.read_sas(
    data_dir / "GHB_G.xpt",
    format="xport"
)

ghb_h = pd.read_sas(
    data_dir / "GHB_H.xpt",
    format="xport"
)

ghb_i = pd.read_sas(
    data_dir / "GHB_I.xpt",
    format="xport"
)

ghb_j = pd.read_sas(
    data_dir / "GHB_J.xpt",
    format="xport"
)

print("2011-2012:", ghb_g.shape)
print("2013-2014:", ghb_h.shape)
print("2015-2016:", ghb_i.shape)
print("2017-2018:", ghb_j.shape)



2011-2012: (6549, 2)
2013-2014: (6979, 2)
2015-2016: (6744, 2)
2017-2018: (6401, 2)


In [20]:
print("2011-2012 columns:")
print(ghb_g.columns.tolist())

print("\n2013-2014 columns:")
print(ghb_h.columns.tolist())

print("\n2015-2016 columns:")
print(ghb_i.columns.tolist())

print("\n2017-2018 columns:")
print(ghb_j.columns.tolist())

2011-2012 columns:
['SEQN', 'LBXGH']

2013-2014 columns:
['SEQN', 'LBXGH']

2015-2016 columns:
['SEQN', 'LBXGH']

2017-2018 columns:
['SEQN', 'LBXGH']


 'SEQN', 'LBXGH' are common in all cycle.

#### Checking how many valid HbA1c measurement each cycle has

In [28]:
ghb_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],
    "total_rows": [
        len(ghb_g),
        len(ghb_h),
        len(ghb_i),
        len(ghb_j)
    ],
    "valid_hba1c": [
        ghb_g["LBXGH"].notna().sum(),
        ghb_h["LBXGH"].notna().sum(),
        ghb_i["LBXGH"].notna().sum(),
        ghb_j["LBXGH"].notna().sum()
    ]
})

ghb_valid_counts

,cycle,total_rows,valid_hba1c
0,2011-2012,6549,6145
1,2013-2014,6979,6643
2,2015-2016,6744,6326
3,2017-2018,6401,6045


In [30]:
print(
    "Total valid HbA1c measurements:",
    ghb_valid_counts["valid_hba1c"].sum()
)

Total valid HbA1c measurements: 25159


### C. Body Measures Data Compatibility

The Body Measures (BMX) component contains physical measurements collected during the NHANES examination.

From data we find that a total of 19 body-measure variables are common across 2011–2018. The three variables ( BMXBMI: Body Mass Index, BMXWAIST: Waist circumference, and BMXWT: Body weight ) planned for machine learning are avialable all cycle thus BMI, waist circumference, and body weight can be used consistently as predictors across the 2011–2018 study period.

Some additional body-measure variables differ between cycles but these variables are not required for analysis right now.

In [32]:
# Load Body Measures datasets 

bmx_g = pd.read_sas(
    data_dir / "BMX_G.xpt",
    format="xport"
)

bmx_h = pd.read_sas(
    data_dir / "BMX_H.xpt",
    format="xport"
)

bmx_i = pd.read_sas(
    data_dir / "BMX_I.xpt",
    format="xport"
)

bmx_j = pd.read_sas(
    data_dir / "BMX_J.xpt",
    format="xport"
)

print("2011-2012:", bmx_g.shape)
print("2013-2014:", bmx_h.shape)
print("2015-2016:", bmx_i.shape)
print("2017-2018:", bmx_j.shape)

2011-2012: (9338, 26)
2013-2014: (9813, 26)
2015-2016: (9544, 26)
2017-2018: (8704, 21)


In [34]:
print("2011-2012 columns:")
print(bmx_g.columns.tolist())

print("\n2013-2014 columns:")
print(bmx_h.columns.tolist())

print("\n2015-2016 columns:")
print(bmx_i.columns.tolist())

print("\n2017-2018 columns:")
print(bmx_j.columns.tolist())

2011-2012 columns:
['SEQN', 'BMDSTATS', 'BMXWT', 'BMIWT', 'BMXRECUM', 'BMIRECUM', 'BMXHEAD', 'BMIHEAD', 'BMXHT', 'BMIHT', 'BMXBMI', 'BMDBMIC', 'BMXLEG', 'BMILEG', 'BMXARML', 'BMIARML', 'BMXARMC', 'BMIARMC', 'BMXWAIST', 'BMIWAIST', 'BMXSAD1', 'BMXSAD2', 'BMXSAD3', 'BMXSAD4', 'BMDAVSAD', 'BMDSADCM']

2013-2014 columns:
['SEQN', 'BMDSTATS', 'BMXWT', 'BMIWT', 'BMXRECUM', 'BMIRECUM', 'BMXHEAD', 'BMIHEAD', 'BMXHT', 'BMIHT', 'BMXBMI', 'BMDBMIC', 'BMXLEG', 'BMILEG', 'BMXARML', 'BMIARML', 'BMXARMC', 'BMIARMC', 'BMXWAIST', 'BMIWAIST', 'BMXSAD1', 'BMXSAD2', 'BMXSAD3', 'BMXSAD4', 'BMDAVSAD', 'BMDSADCM']

2015-2016 columns:
['SEQN', 'BMDSTATS', 'BMXWT', 'BMIWT', 'BMXRECUM', 'BMIRECUM', 'BMXHEAD', 'BMIHEAD', 'BMXHT', 'BMIHT', 'BMXBMI', 'BMDBMIC', 'BMXLEG', 'BMILEG', 'BMXARML', 'BMIARML', 'BMXARMC', 'BMIARMC', 'BMXWAIST', 'BMIWAIST', 'BMXSAD1', 'BMXSAD2', 'BMXSAD3', 'BMXSAD4', 'BMDAVSAD', 'BMDSADCM']

2017-2018 columns:
['SEQN', 'BMDSTATS', 'BMXWT', 'BMIWT', 'BMXRECUM', 'BMIRECUM', 'BMXHEAD', 'BMIHEA

In [36]:
common_bmx_columns = (
    set(bmx_g.columns)
    & set(bmx_h.columns)
    & set(bmx_i.columns)
    & set(bmx_j.columns)
)

print("Number of BMX variables common to all four cycles:")
print(len(common_bmx_columns))

print("\nCommon BMX variables:")
print(sorted(common_bmx_columns))

Number of BMX variables common to all four cycles:
19

Common BMX variables:
['BMDSTATS', 'BMIARMC', 'BMIARML', 'BMIHEAD', 'BMIHT', 'BMILEG', 'BMIRECUM', 'BMIWAIST', 'BMIWT', 'BMXARMC', 'BMXARML', 'BMXBMI', 'BMXHEAD', 'BMXHT', 'BMXLEG', 'BMXRECUM', 'BMXWAIST', 'BMXWT', 'SEQN']


In [38]:
candidate_bmx_variables = [
    "SEQN",
    "BMXWT",
    "BMXBMI",
    "BMXWAIST"
]

bmx_datasets = {
    "2011-2012": bmx_g,
    "2013-2014": bmx_h,
    "2015-2016": bmx_i,
    "2017-2018": bmx_j
}

bmx_compatibility = []

for variable in candidate_bmx_variables:

    row = {"variable": variable}

    for cycle, df in bmx_datasets.items():
        row[cycle] = variable in df.columns

    bmx_compatibility.append(row)

bmx_compatibility_df = pd.DataFrame(bmx_compatibility)

bmx_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,BMXWT,True,True,True,True
2,BMXBMI,True,True,True,True
3,BMXWAIST,True,True,True,True


In [40]:
bmx_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],

    "total_rows": [
        len(bmx_g),
        len(bmx_h),
        len(bmx_i),
        len(bmx_j)
    ],

    "valid_bmi": [
        bmx_g["BMXBMI"].notna().sum(),
        bmx_h["BMXBMI"].notna().sum(),
        bmx_i["BMXBMI"].notna().sum(),
        bmx_j["BMXBMI"].notna().sum()
    ],

    "valid_waist": [
        bmx_g["BMXWAIST"].notna().sum(),
        bmx_h["BMXWAIST"].notna().sum(),
        bmx_i["BMXWAIST"].notna().sum(),
        bmx_j["BMXWAIST"].notna().sum()
    ],

    "valid_weight": [
        bmx_g["BMXWT"].notna().sum(),
        bmx_h["BMXWT"].notna().sum(),
        bmx_i["BMXWT"].notna().sum(),
        bmx_j["BMXWT"].notna().sum()
    ]
})

bmx_valid_counts

,cycle,total_rows,valid_bmi,valid_waist,valid_weight
0,2011-2012,9338,8602,8204,9243
1,2013-2014,9813,9055,8661,9723
2,2015-2016,9544,8756,8313,9445
3,2017-2018,8704,8005,7601,8580


In [42]:
print(
    "Total valid Body Mass Index:",
    bmx_valid_counts["valid_bmi"].sum()
)
print(
    "Total valid Body waist measurements:",
    bmx_valid_counts["valid_waist"].sum()
)
print(
    "Total valid Body weight measurements:",
    bmx_valid_counts["valid_weight"].sum()
)

Total valid Body Mass Index: 34418
Total valid Body waist measurements: 32779
Total valid Body weight measurements: 36991


### D. Diabetes Questionnaire Compatibility

The Diabetes Questionnaire (DIQ) contains self-reported information about diabetes, prediabetes, treatment, and related health history. It has a total of 53 variables that are common across 2011–2018. I wanted to look at DIQ010 for this project and it is avialable across the four cycle.  DIQ010 records whether a participant reports having been told by a doctor or health professional that they have diabetes, excluding diabetes that occurred only during pregnancy and is coded as below.

1 = Yes
2 = No
3 = Borderline
7 = Refused
9 = Don’t know
missing = no recorded response

For the Machine learning I will be excluding DIQ010 = 1 as it will be treated as having a previously reported diabetes diagnosis and treat 7 or 9 as missing or unknown, 3 will be examined during data preparation before the final eligibility rule is established.

DIQ010 = 1   known diabetes
DIQ010 = 3   told borderline
DIQ010 = 2   No known diabetes

In [44]:
# Load Diabetes Questionnaire datasets 

diq_g = pd.read_sas(
    data_dir / "DIQ_G.xpt",
    format="xport"
)

diq_h = pd.read_sas(
    data_dir / "DIQ_H.xpt",
    format="xport"
)

diq_i = pd.read_sas(
    data_dir / "DIQ_I.xpt",
    format="xport"
)

diq_j = pd.read_sas(
    data_dir / "DIQ_J.xpt",
    format="xport"
)

print("2011-2012:", diq_g.shape)
print("2013-2014:", diq_h.shape)
print("2015-2016:", diq_i.shape)
print("2017-2018:", diq_j.shape)

2011-2012: (9364, 53)
2013-2014: (9770, 54)
2015-2016: (9575, 54)
2017-2018: (8897, 54)


In [46]:
print("2011-2012 columns:")
print(diq_g.columns.tolist())

print("\n2013-2014 columns:")
print(diq_h.columns.tolist())

print("\n2015-2016 columns:")
print(diq_i.columns.tolist())

print("\n2017-2018 columns:")
print(diq_j.columns.tolist())

2011-2012 columns:
['SEQN', 'DIQ010', 'DID040', 'DIQ160', 'DIQ170', 'DIQ172', 'DIQ175A', 'DIQ175B', 'DIQ175C', 'DIQ175D', 'DIQ175E', 'DIQ175F', 'DIQ175G', 'DIQ175H', 'DIQ175I', 'DIQ175J', 'DIQ175K', 'DIQ175L', 'DIQ175M', 'DIQ175N', 'DIQ175O', 'DIQ175P', 'DIQ175Q', 'DIQ175R', 'DIQ175S', 'DIQ175T', 'DIQ175U', 'DIQ175V', 'DIQ175W', 'DIQ180', 'DIQ050', 'DID060', 'DIQ060U', 'DIQ070', 'DIQ230', 'DIQ240', 'DID250', 'DID260', 'DIQ260U', 'DIQ275', 'DIQ280', 'DIQ291', 'DIQ300S', 'DIQ300D', 'DID310S', 'DID310D', 'DID320', 'DID330', 'DID341', 'DID350', 'DIQ350U', 'DIQ360', 'DIQ080']

2013-2014 columns:
['SEQN', 'DIQ010', 'DID040', 'DIQ160', 'DIQ170', 'DIQ172', 'DIQ175A', 'DIQ175B', 'DIQ175C', 'DIQ175D', 'DIQ175E', 'DIQ175F', 'DIQ175G', 'DIQ175H', 'DIQ175I', 'DIQ175J', 'DIQ175K', 'DIQ175L', 'DIQ175M', 'DIQ175N', 'DIQ175O', 'DIQ175P', 'DIQ175Q', 'DIQ175R', 'DIQ175S', 'DIQ175T', 'DIQ175U', 'DIQ175V', 'DIQ175W', 'DIQ175X', 'DIQ180', 'DIQ050', 'DID060', 'DIQ060U', 'DIQ070', 'DIQ230', 'DIQ240', 'DID250'

In [48]:
common_diq_columns = (
    set(diq_g.columns)
    & set(diq_h.columns)
    & set(diq_i.columns)
    & set(diq_j.columns)
)

print("Number of DIQ variables common to all four cycles:")
print(len(common_diq_columns))

print("\nCommon DIQ variables:")
print(sorted(common_diq_columns))

Number of DIQ variables common to all four cycles:
53

Common DIQ variables:
['DID040', 'DID060', 'DID250', 'DID260', 'DID310D', 'DID310S', 'DID320', 'DID330', 'DID341', 'DID350', 'DIQ010', 'DIQ050', 'DIQ060U', 'DIQ070', 'DIQ080', 'DIQ160', 'DIQ170', 'DIQ172', 'DIQ175A', 'DIQ175B', 'DIQ175C', 'DIQ175D', 'DIQ175E', 'DIQ175F', 'DIQ175G', 'DIQ175H', 'DIQ175I', 'DIQ175J', 'DIQ175K', 'DIQ175L', 'DIQ175M', 'DIQ175N', 'DIQ175O', 'DIQ175P', 'DIQ175Q', 'DIQ175R', 'DIQ175S', 'DIQ175T', 'DIQ175U', 'DIQ175V', 'DIQ175W', 'DIQ180', 'DIQ230', 'DIQ240', 'DIQ260U', 'DIQ275', 'DIQ280', 'DIQ291', 'DIQ300D', 'DIQ300S', 'DIQ350U', 'DIQ360', 'SEQN']


In [50]:
#the variables that I want to look into
candidate_diq_variables = [
    "SEQN",
    "DIQ010"
]

diq_datasets = {
    "2011-2012": diq_g,
    "2013-2014": diq_h,
    "2015-2016": diq_i,
    "2017-2018": diq_j
}

diq_compatibility = []

for variable in candidate_diq_variables:

    row = {"variable": variable}

    for cycle, df in diq_datasets.items():
        row[cycle] = variable in df.columns

    diq_compatibility.append(row)

diq_compatibility_df = pd.DataFrame(diq_compatibility)

diq_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,DIQ010,True,True,True,True


In [56]:
diq_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],

    "total_rows": [
        len(diq_g),
        len(diq_h),
        len(diq_i),
        len(diq_j)
    ],

    "valid_diq010": [
        diq_g["DIQ010"].notna().sum(),
        diq_h["DIQ010"].notna().sum(),
        diq_i["DIQ010"].notna().sum(),
        diq_j["DIQ010"].notna().sum()
    ]
})
#valid_diq010 = Doctor told you have diabetes.
diq_valid_counts

,cycle,total_rows,valid_diq010
0,2011-2012,9364,9363
1,2013-2014,9770,9769
2,2015-2016,9575,9575
3,2017-2018,8897,8897


In [58]:
print(
    "Total valid DIQ010 responses:",
    diq_valid_counts["valid_diq010"].sum()
)

Total valid DIQ010 responses: 37604


In [60]:
diq_g["DIQ010"].value_counts(dropna=False).sort_index()

DIQ010
1.0     708
2.0    8524
3.0     125
7.0       1
9.0       5
NaN       1
Name: count, dtype: int64

**Diabetese diagonisis** :
708   diagnosed diabetes,
8524  no diagnosed diabetes,
125   borderline,
1     refused,
5     don't know,
1     missing

### E. Physical Activity

The Physical Activity (PAQ) component contains questionnaire information about participant's physical activity and sedentary behavior.

From the data, we see that a total of 17 physical-activity variables are common across all four cycles.
The variable PAD680, which represents sedentary time in minutes per day is available in all four cycles and contains a large number of usable observations.

During exploration, special values such as 7777 and 9999 were identified, it represent non-substantive responses such as refusal or "don't know" and will be converted to missing values before statistical analysis or machine-learning modeling.

 small floating-point values close to zero may also appear when SAS Transport files are read into pandas. These will be treated as zero where appropriate after verifying the original NHANES coding.

#### Planned Future Use

Physical-activity variables will not be used in their raw form immediately. Their definitions and units will first be confirmed using the official NHANES codebooks. I might create harmonized behavioral features such as sedentary_minutes_per_day, moderate_activity_minutes_per_week, and vigorous_activity_minutes_per_week 

These engineered variables may be included as predictors in later machine- learning models to determine whether physical activity and sedentary behavior provide additional information for identifying elevated HbA1c beyond traditional factors such as age and BMI.

The final decision about which physical-activity variables to retain will be
made after reviewing missing values, response coding, and compatibility across
all four survey cycles.
 

In [67]:
# Load Physical Activity datasets 

paq_g = pd.read_sas(
    data_dir / "PAQ_G.xpt",
    format="xport"
)

paq_h = pd.read_sas(
    data_dir / "PAQ_H.xpt",
    format="xport"
)

paq_i = pd.read_sas(
    data_dir / "PAQ_I.xpt",
    format="xport"
)

paq_j = pd.read_sas(
    data_dir / "PAQ_J.xpt",
    format="xport"
)
# Compare the number of rows and columns in each cycle

print("2011-2012:", paq_g.shape)
print("2013-2014:", paq_h.shape)
print("2015-2016:", paq_i.shape)
print("2017-2018:", paq_j.shape)

2011-2012: (9107, 21)
2013-2014: (9484, 96)
2015-2016: (9255, 94)
2017-2018: (5856, 17)


In [69]:

print("2011-2012 columns:")
print(paq_g.columns.tolist())

print("\n2013-2014 columns:")
print(paq_h.columns.tolist())

print("\n2015-2016 columns:")
print(paq_i.columns.tolist())

print("\n2017-2018 columns:")
print(paq_j.columns.tolist())

2011-2012 columns:
['SEQN', 'PAQ706', 'PAQ605', 'PAQ610', 'PAD615', 'PAQ620', 'PAQ625', 'PAD630', 'PAQ635', 'PAQ640', 'PAD645', 'PAQ650', 'PAQ655', 'PAD660', 'PAQ665', 'PAQ670', 'PAD675', 'PAD680', 'PAQ710', 'PAQ715', 'PAAQUEX']

2013-2014 columns:
['SEQN', 'PAQ605', 'PAQ610', 'PAD615', 'PAQ620', 'PAQ625', 'PAD630', 'PAQ635', 'PAQ640', 'PAD645', 'PAQ650', 'PAQ655', 'PAD660', 'PAQ665', 'PAQ670', 'PAD675', 'PAD680', 'PAQ706', 'PAQ710', 'PAQ715', 'PAQ722', 'PAQ724A', 'PAQ724B', 'PAQ724C', 'PAQ724D', 'PAQ724E', 'PAQ724F', 'PAQ724G', 'PAQ724H', 'PAQ724I', 'PAQ724J', 'PAQ724K', 'PAQ724L', 'PAQ724M', 'PAQ724N', 'PAQ724O', 'PAQ724P', 'PAQ724Q', 'PAQ724R', 'PAQ724S', 'PAQ724T', 'PAQ724U', 'PAQ724V', 'PAQ724W', 'PAQ724X', 'PAQ724Y', 'PAQ724Z', 'PAQ724AA', 'PAQ724AB', 'PAQ724AC', 'PAQ724AD', 'PAQ724AE', 'PAQ724AF', 'PAQ724CM', 'PAQ731', 'PAD733', 'PAQ677', 'PAQ678', 'PAQ740', 'PAQ742', 'PAQ744', 'PAQ746', 'PAQ748', 'PAQ755', 'PAQ759A', 'PAQ759B', 'PAQ759C', 'PAQ759D', 'PAQ759E', 'PAQ759F', 'PAQ75

In [75]:
# intersection of the four sets of column names

common_paq_columns = (
    set(paq_g.columns)
    & set(paq_h.columns)
    & set(paq_i.columns)
    & set(paq_j.columns)
)

print("Number of PAQ variables common to all four cycles:")
print(len(common_paq_columns))

print("\nCommon PAQ variables:")
print(sorted(common_paq_columns))

Number of PAQ variables common to all four cycles:
17

Common PAQ variables:
['PAD615', 'PAD630', 'PAD645', 'PAD660', 'PAD675', 'PAD680', 'PAQ605', 'PAQ610', 'PAQ620', 'PAQ625', 'PAQ635', 'PAQ640', 'PAQ650', 'PAQ655', 'PAQ665', 'PAQ670', 'SEQN']


In [77]:

all_paq_columns = (
    set(paq_g.columns)
    | set(paq_h.columns)
    | set(paq_i.columns)
    | set(paq_j.columns)
)

missing_from_g = all_paq_columns - set(paq_g.columns)
missing_from_h = all_paq_columns - set(paq_h.columns)
missing_from_i = all_paq_columns - set(paq_i.columns)
missing_from_j = all_paq_columns - set(paq_j.columns)

print("Missing from 2011-2012:")
print(sorted(missing_from_g))

print("\nMissing from 2013-2014:")
print(sorted(missing_from_h))

print("\nMissing from 2015-2016:")
print(sorted(missing_from_i))

print("\nMissing from 2017-2018:")
print(sorted(missing_from_j))

Missing from 2011-2012:
['PAD733', 'PAQ677', 'PAQ678', 'PAQ679', 'PAQ722', 'PAQ724A', 'PAQ724AA', 'PAQ724AB', 'PAQ724AC', 'PAQ724AD', 'PAQ724AE', 'PAQ724AF', 'PAQ724B', 'PAQ724C', 'PAQ724CM', 'PAQ724D', 'PAQ724E', 'PAQ724F', 'PAQ724G', 'PAQ724H', 'PAQ724I', 'PAQ724J', 'PAQ724K', 'PAQ724L', 'PAQ724M', 'PAQ724N', 'PAQ724O', 'PAQ724P', 'PAQ724Q', 'PAQ724R', 'PAQ724S', 'PAQ724T', 'PAQ724U', 'PAQ724V', 'PAQ724W', 'PAQ724X', 'PAQ724Y', 'PAQ724Z', 'PAQ731', 'PAQ740', 'PAQ742', 'PAQ744', 'PAQ746', 'PAQ748', 'PAQ750', 'PAQ755', 'PAQ759A', 'PAQ759B', 'PAQ759C', 'PAQ759D', 'PAQ759E', 'PAQ759F', 'PAQ759G', 'PAQ759H', 'PAQ759I', 'PAQ759J', 'PAQ759K', 'PAQ759L', 'PAQ759M', 'PAQ759N', 'PAQ759O', 'PAQ759P', 'PAQ759Q', 'PAQ759R', 'PAQ759S', 'PAQ759T', 'PAQ759U', 'PAQ759V', 'PAQ762', 'PAQ764', 'PAQ766', 'PAQ770', 'PAQ772A', 'PAQ772B', 'PAQ772C']

Missing from 2013-2014:
[]

Missing from 2015-2016:
['PAAQUEX', 'PAQ759V']

Missing from 2017-2018:
['PAAQUEX', 'PAD733', 'PAQ677', 'PAQ678', 'PAQ679', 'PAQ706

In [83]:
# Checking whether SEQN and PAD680 exist in all four cycles

candidate_paq_variables = [
    "SEQN",
    "PAD680"
]

paq_datasets = {
    "2011-2012": paq_g,
    "2013-2014": paq_h,
    "2015-2016": paq_i,
    "2017-2018": paq_j
}

paq_compatibility = []

for variable in candidate_paq_variables:

    row = {"variable": variable}

    for cycle, df in paq_datasets.items():
        row[cycle] = variable in df.columns

    paq_compatibility.append(row)

paq_compatibility_df = pd.DataFrame(paq_compatibility)

paq_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,PAD680,True,True,True,True


In [85]:
#  non-missing PAD680 observations in each cycle

paq_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],

    "total_rows": [
        len(paq_g),
        len(paq_h),
        len(paq_i),
        len(paq_j)
    ],

    "valid_pad680": [
        paq_g["PAD680"].notna().sum(),
        paq_h["PAD680"].notna().sum(),
        paq_i["PAD680"].notna().sum(),
        paq_j["PAD680"].notna().sum()
    ]
})
#sedentary time in minutes per day 
paq_valid_counts

,cycle,total_rows,valid_pad680
0,2011-2012,9107,6773
1,2013-2014,9484,7139
2,2015-2016,9255,6951
3,2017-2018,5856,5846


In [87]:
print(
    "Total valid sedentary-time responses:",
    paq_valid_counts["valid_pad680"].sum()
)

Total valid sedentary-time responses: 26709


In [89]:
# Examining the values recorded in PAD680 for one cycle

paq_g["PAD680"].value_counts(
    dropna=False
).sort_index()

PAD680
5.397605e-79       1
3.000000e+00       4
4.000000e+00       2
5.000000e+00       2
6.000000e+00       2
7.000000e+00       1
8.000000e+00       5
1.000000e+01       5
1.100000e+01       1
1.300000e+01       1
1.500000e+01       4
1.600000e+01       1
2.000000e+01       8
3.000000e+01      33
4.000000e+01       3
4.500000e+01       4
5.500000e+01       1
6.000000e+01     231
9.000000e+01      15
1.200000e+02     539
1.500000e+02       3
1.800000e+02     661
2.100000e+02       1
2.400000e+02     777
3.000000e+02     650
3.600000e+02     731
4.200000e+02     350
4.300000e+02       1
4.800000e+02    1055
5.400000e+02     296
6.000000e+02     634
6.600000e+02     117
7.200000e+02     367
7.800000e+02      64
8.400000e+02      75
9.000000e+02      41
9.600000e+02      39
1.020000e+03       5
1.080000e+03       6
1.140000e+03       3
1.200000e+03       2
1.380000e+03       1
7.777000e+03       1
9.999000e+03      30
NaN             2334
Name: count, dtype: int64

In [91]:
for cycle, df in paq_datasets.items():

    print(f"\n{cycle}")
    print("-" * 30)

    print(df["PAD680"].describe())


2011-2012
------------------------------
count    6.773000e+03
mean     4.274666e+02
std      6.758443e+02
min      5.397605e-79
25%      2.400000e+02
50%      3.600000e+02
75%      4.800000e+02
max      9.999000e+03
Name: PAD680, dtype: float64

2013-2014
------------------------------
count    7.139000e+03
mean     4.785490e+02
std      6.443079e+02
min      5.397605e-79
25%      3.000000e+02
50%      4.800000e+02
75%      6.000000e+02
max      9.999000e+03
Name: PAD680, dtype: float64

2015-2016
------------------------------
count    6.951000e+03
mean     4.684020e+02
std      8.725388e+02
min      5.397605e-79
25%      2.400000e+02
50%      3.600000e+02
75%      5.400000e+02
max      9.999000e+03
Name: PAD680, dtype: float64

2017-2018
------------------------------
count    5.846000e+03
mean     3.895582e+02
std      7.718131e+02
min      5.397605e-79
25%      1.800000e+02
50%      3.000000e+02
75%      4.800000e+02
max      9.999000e+03
Name: PAD680, dtype: float64


### F. Smoking Data Compatibility

The Smoking (SMQ) component contains information about lifetime cigarette use and current smoking behavior. A total of 22 smoking-related variables are common across 2011–2018. The two smoking variables I want to use SMQ020 (whether the participant has smoked at least 100 cigarettes in their lifetime), SMQ040 (current smoking status )for this project are available in every cycle:



In [93]:
# Load Smoking datasets 

smq_g = pd.read_sas(
    data_dir / "SMQ_G.xpt",
    format="xport"
)

smq_h = pd.read_sas(
    data_dir / "SMQ_H.xpt",
    format="xport"
)

smq_i = pd.read_sas(
    data_dir / "SMQ_I.xpt",
    format="xport"
)

smq_j = pd.read_sas(
    data_dir / "SMQ_J.xpt",
    format="xport"
)


print("2011-2012:", smq_g.shape)
print("2013-2014:", smq_h.shape)
print("2015-2016:", smq_i.shape)
print("2017-2018:", smq_j.shape)

2011-2012: (6790, 30)
2013-2014: (7168, 32)
2015-2016: (7001, 42)
2017-2018: (6724, 37)


In [95]:

common_smq_columns = (
    set(smq_g.columns)
    & set(smq_h.columns)
    & set(smq_i.columns)
    & set(smq_j.columns)
)

print("Number of SMQ variables common to all four cycles:")
print(len(common_smq_columns))

print("\nCommon SMQ variables:")
print(sorted(common_smq_columns))

Number of SMQ variables common to all four cycles:
22

Common SMQ variables:
['SEQN', 'SMAQUEX2', 'SMD030', 'SMD057', 'SMD093', 'SMD100BR', 'SMD100CO', 'SMD100FL', 'SMD100LN', 'SMD100MN', 'SMD100NI', 'SMD100TR', 'SMD630', 'SMD641', 'SMD650', 'SMDUPCA', 'SMQ020', 'SMQ040', 'SMQ050Q', 'SMQ050U', 'SMQ621', 'SMQ670']


In [97]:

candidate_smq_variables = [
    "SEQN",
    "SMQ020",
    "SMQ040"
]

smq_datasets = {
    "2011-2012": smq_g,
    "2013-2014": smq_h,
    "2015-2016": smq_i,
    "2017-2018": smq_j
}

smq_compatibility = []

for variable in candidate_smq_variables:

    row = {"variable": variable}

    for cycle, df in smq_datasets.items():
        row[cycle] = variable in df.columns

    smq_compatibility.append(row)

smq_compatibility_df = pd.DataFrame(smq_compatibility)

smq_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,SMQ020,True,True,True,True
2,SMQ040,True,True,True,True


In [99]:
# Count non-missing smoking responses by cycle

smq_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],

    "total_rows": [
        len(smq_g),
        len(smq_h),
        len(smq_i),
        len(smq_j)
    ],

    "valid_smq020": [
        smq_g["SMQ020"].notna().sum(),
        smq_h["SMQ020"].notna().sum(),
        smq_i["SMQ020"].notna().sum(),
        smq_j["SMQ020"].notna().sum()
    ],

    "valid_smq040": [
        smq_g["SMQ040"].notna().sum(),
        smq_h["SMQ040"].notna().sum(),
        smq_i["SMQ040"].notna().sum(),
        smq_j["SMQ040"].notna().sum()
    ]
})

smq_valid_counts

,cycle,total_rows,valid_smq020,valid_smq040
0,2011-2012,6790,5559,2369
1,2013-2014,7168,6113,2579
2,2015-2016,7001,5992,2422
3,2017-2018,6724,5856,2359


In [101]:
# response code 

print("SMQ020 values:")
print(
    smq_g["SMQ020"]
    .value_counts(dropna=False)
    .sort_index()
)

print("\nSMQ040 values:")
print(
    smq_g["SMQ040"]
    .value_counts(dropna=False)
    .sort_index()
)

SMQ020 values:
SMQ020
1.0    2369
2.0    3184
7.0       2
9.0       4
NaN    1231
Name: count, dtype: int64

SMQ040 values:
SMQ040
1.0     912
2.0     196
3.0    1259
7.0       2
NaN    4421
Name: count, dtype: int64


### G. Sleep Data Compatibility

The Sleep Disorders (SLQ) component contains information about participant's sleep duration and selected sleep related behaviors. It contain only two variables with the same names across all four survey cycles: SEQN and SLQ050. 

However, the main sleep-duration measure is still available across 2011–2018 under different variable names SLD010H  for 2011–2012 and 2013–2014, SLD012 for 2015–2016 and 2017–2018. when combining the data set it can be put together as sleep_hours. 

Special response codes such as "Refused" and "Don't know" will be converted
to missing values before analysis.

I mighgt use Sleep duration as a behavioral predictor in themachine-learning models.

In [103]:

slq_g = pd.read_sas(
    data_dir / "SLQ_G.xpt",
    format="xport"
)

slq_h = pd.read_sas(
    data_dir / "SLQ_H.xpt",
    format="xport"
)

slq_i = pd.read_sas(
    data_dir / "SLQ_I.xpt",
    format="xport"
)

slq_j = pd.read_sas(
    data_dir / "SLQ_J.xpt",
    format="xport"
)
print("2011-2012:", slq_g.shape)
print("2013-2014:", slq_h.shape)
print("2015-2016:", slq_i.shape)
print("2017-2018:", slq_j.shape)


2011-2012: (6175, 4)
2013-2014: (6464, 4)
2015-2016: (6327, 8)
2017-2018: (6161, 11)


In [105]:
print("2011-2012 columns:")
print(slq_g.columns.tolist())

print("\n2013-2014 columns:")
print(slq_h.columns.tolist())

print("\n2015-2016 columns:")
print(slq_i.columns.tolist())

print("\n2017-2018 columns:")
print(slq_j.columns.tolist())

2011-2012 columns:
['SEQN', 'SLD010H', 'SLQ050', 'SLQ060']

2013-2014 columns:
['SEQN', 'SLD010H', 'SLQ050', 'SLQ060']

2015-2016 columns:
['SEQN', 'SLQ300', 'SLQ310', 'SLD012', 'SLQ030', 'SLQ040', 'SLQ050', 'SLQ120']

2017-2018 columns:
['SEQN', 'SLQ300', 'SLQ310', 'SLD012', 'SLQ320', 'SLQ330', 'SLD013', 'SLQ030', 'SLQ040', 'SLQ050', 'SLQ120']


In [107]:
common_slq_columns = (
    set(slq_g.columns)
    & set(slq_h.columns)
    & set(slq_i.columns)
    & set(slq_j.columns)
)

print("Number of SLQ variables common to all four cycles:")
print(len(common_slq_columns))

print("\nCommon SLQ variables:")
print(sorted(common_slq_columns))

Number of SLQ variables common to all four cycles:
2

Common SLQ variables:
['SEQN', 'SLQ050']


In [109]:
possible_sleep_variables = [
    "SLD010H",
    "SLD012",
    "SLQ300",
    "SLQ310"
]

slq_datasets = {
    "2011-2012": slq_g,
    "2013-2014": slq_h,
    "2015-2016": slq_i,
    "2017-2018": slq_j
}

sleep_compatibility = []

for variable in possible_sleep_variables:

    row = {"variable": variable}

    for cycle, df in slq_datasets.items():
        row[cycle] = variable in df.columns

    sleep_compatibility.append(row)

sleep_compatibility_df = pd.DataFrame(sleep_compatibility)

sleep_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SLD010H,True,True,False,False
1,SLD012,False,False,True,True
2,SLQ300,False,False,True,True
3,SLQ310,False,False,True,True


In [111]:
for cycle, df in slq_datasets.items():

    print(f"\n{cycle}")
    print("-" * 30)

    if "SLD010H" in df.columns:
        print(df["SLD010H"].describe())

    if "SLD012" in df.columns:
        print(df["SLD012"].describe())


2011-2012
------------------------------
count    6166.000000
mean        7.005352
std         3.527228
min         2.000000
25%         6.000000
50%         7.000000
75%         8.000000
max        99.000000
Name: SLD010H, dtype: float64

2013-2014
------------------------------
count    6461.000000
mean        7.050302
std         3.347950
min         2.000000
25%         6.000000
50%         7.000000
75%         8.000000
max        99.000000
Name: SLD010H, dtype: float64

2015-2016
------------------------------
count    6294.000000
mean        7.752780
std         1.568593
min         2.000000
25%         7.000000
50%         8.000000
75%         8.500000
max        14.500000
Name: SLD012, dtype: float64

2017-2018
------------------------------
count    6113.000000
mean        7.658842
std         1.669706
min         2.000000
25%         7.000000
50%         8.000000
75%         8.500000
max        14.000000
Name: SLD012, dtype: float64


### H. Blood Pressure & Cholesterol Questionnaire

The Blood Pressure & Cholesterol Questionnaire (BPQ) contains self-reported information about hypertension and high cholesterol. A total of 11 BPQ variables are common across 2011–2018. 

BPQ020 : whether the participant has ever been told they have high blood pressure or hypertension

BPQ080 : whether the participant has ever been told they have high cholesterol

These both variables have a large number of valid responses across the four cycles, and I might  use it as cardiovascular risk predictors in the machine-learning model.

Special response codes such as `7` (Refused) and `9` (Don't know) will be
treated as missing values during data cleaning.





In [113]:
# Load Blood Pressure & Cholesterol Questionnaire datasets

bpq_g = pd.read_sas(
    data_dir / "BPQ_G.xpt",
    format="xport"
)

bpq_h = pd.read_sas(
    data_dir / "BPQ_H.xpt",
    format="xport"
)

bpq_i = pd.read_sas(
    data_dir / "BPQ_I.xpt",
    format="xport"
)

bpq_j = pd.read_sas(
    data_dir / "BPQ_J.xpt",
    format="xport"
)

print("2011-2012:", bpq_g.shape)
print("2013-2014:", bpq_h.shape)
print("2015-2016:", bpq_i.shape)
print("2017-2018:", bpq_j.shape)

2011-2012: (6175, 15)
2013-2014: (6464, 14)
2015-2016: (6327, 11)
2017-2018: (6161, 11)


In [115]:

common_bpq_columns = (
    set(bpq_g.columns)
    & set(bpq_h.columns)
    & set(bpq_i.columns)
    & set(bpq_j.columns)
)

print("Number of BPQ variables common to all four cycles:")
print(len(common_bpq_columns))

print("\nCommon BPQ variables:")
print(sorted(common_bpq_columns))

Number of BPQ variables common to all four cycles:
11

Common BPQ variables:
['BPD035', 'BPQ020', 'BPQ030', 'BPQ040A', 'BPQ050A', 'BPQ060', 'BPQ070', 'BPQ080', 'BPQ090D', 'BPQ100D', 'SEQN']


In [117]:
# variables intreasted in
candidate_bpq_variables = [
    "SEQN",
    "BPQ020",   # High blood pressure / hypertension
    "BPQ080"    # High cholesterol
]

bpq_datasets = {
    "2011-2012": bpq_g,
    "2013-2014": bpq_h,
    "2015-2016": bpq_i,
    "2017-2018": bpq_j
}

bpq_compatibility = []

for variable in candidate_bpq_variables:

    row = {"variable": variable}

    for cycle, df in bpq_datasets.items():
        row[cycle] = variable in df.columns

    bpq_compatibility.append(row)

bpq_compatibility_df = pd.DataFrame(bpq_compatibility)

bpq_compatibility_df

,variable,2011-2012,2013-2014,2015-2016,2017-2018
0,SEQN,True,True,True,True
1,BPQ020,True,True,True,True
2,BPQ080,True,True,True,True


In [121]:
# Counting non-missing responses for hypertension and high cholesterol

bpq_valid_counts = pd.DataFrame({
    "cycle": [
        "2011-2012",
        "2013-2014",
        "2015-2016",
        "2017-2018"
    ],

    "total_rows": [
        len(bpq_g),
        len(bpq_h),
        len(bpq_i),
        len(bpq_j)
    ],

    "valid_bpq020": [
        bpq_g["BPQ020"].notna().sum(),
        bpq_h["BPQ020"].notna().sum(),
        bpq_i["BPQ020"].notna().sum(),
        bpq_j["BPQ020"].notna().sum()
    ],

    "valid_bpq080": [
        bpq_g["BPQ080"].notna().sum(),
        bpq_h["BPQ080"].notna().sum(),
        bpq_i["BPQ080"].notna().sum(),
        bpq_j["BPQ080"].notna().sum()
    ]
})

bpq_valid_counts

,cycle,total_rows,valid_bpq020,valid_bpq080
0,2011-2012,6175,6174,5559
1,2013-2014,6464,6464,6464
2,2015-2016,6327,6327,6327
3,2017-2018,6161,6161,6161


In [123]:
print("BPQ020 values:")
print(
    bpq_g["BPQ020"]
    .value_counts(dropna=False)
    .sort_index()
)

print("\nBPQ080 values:")
print(
    bpq_g["BPQ080"]
    .value_counts(dropna=False)
    .sort_index()
)

BPQ020 values:
BPQ020
1.0    2031
2.0    4134
9.0       9
NaN       1
Name: count, dtype: int64

BPQ080 values:
BPQ080
1.0    1826
2.0    3696
7.0       1
9.0      36
NaN     616
Name: count, dtype: int64


# 4. Selected Variables for Analysis

After reviewing variable availability and compatibility across the four NHANES
survey cycles, I selected the variables that are most relevant to the research
questions.

| Component | Variable | Description | Role in Project | Planned Final Name |
|---|---|---|---|---|
| DEMO | `SEQN` | Participant identifier | Merge key / ID | `SEQN` |
| DEMO | `RIDAGEYR` | Age in years | Predictor / eligibility | `age` |
| DEMO | `RIAGENDR` | Sex | Predictor | `sex` |
| DEMO | `RIDRETH3` | Race/ethnicity | Predictor | `race_ethnicity` |
| DEMO | `DMDEDUC2` | Education level for adults | Predictor | `education` |
| DEMO | `INDFMPIR` | Family income-to-poverty ratio | Predictor | `income_poverty_ratio` |
| DEMO | `WTMEC2YR` | MEC examination survey weight | Survey analysis | `mec_weight` |
| DEMO | `SDMVPSU` | Survey primary sampling unit | Survey analysis | `psu` |
| DEMO | `SDMVSTRA` | Survey stratum | Survey analysis | `stratum` |
| GHB | `LBXGH` | Glycohemoglobin (HbA1c %) | Target source | `hba1c` |
| BMX | `BMXBMI` | Body Mass Index | Predictor | `bmi` |
| BMX | `BMXWAIST` | Waist circumference | Predictor | `waist_cm` |
| BMX | `BMXWT` | Body weight | EDA / possible predictor | `weight_kg` |
| DIQ | `DIQ010` | Self-reported diabetes diagnosis | Eligibility filter | `diabetes_report` |
| PAQ | `PAQ650` | Vigorous recreational activity | Feature engineering | `vigorous_activity` |
| PAQ | `PAQ655` | Days of vigorous recreational activity | Feature engineering | `vigorous_days` |
| PAQ | `PAD660` | Minutes of vigorous recreational activity | Feature engineering | `vigorous_minutes` |
| PAQ | `PAQ665` | Moderate recreational activity | Feature engineering | `moderate_activity` |
| PAQ | `PAQ670` | Days of moderate recreational activity | Feature engineering | `moderate_days` |
| PAQ | `PAD675` | Minutes of moderate recreational activity | Feature engineering | `moderate_minutes` |
| PAQ | `PAD680` | Sedentary minutes per day | Predictor | `sedentary_minutes` |
| SMQ | `SMQ020` | Smoked at least 100 cigarettes in lifetime | Smoking feature | `smoked_100_cigarettes` |
| SMQ | `SMQ040` | Current smoking status | Smoking feature | `current_smoking` |
| SLQ | `SLD010H` / `SLD012` | Usual weekday/workday sleep duration | Predictor | `sleep_hours` |
| BPQ | `BPQ020` | Reported hypertension | Predictor | `hypertension` |
| BPQ | `BPQ080` | Reported high cholesterol | Predictor | `high_cholesterol` |

In [129]:
# Selected raw variables from each NHANES component

selected_variables = {
    
    "DEMO": [
        "SEQN",
        "RIDAGEYR",
        "RIAGENDR",
        "RIDRETH3",
        "DMDEDUC2",
        "INDFMPIR",
        "WTMEC2YR",
        "SDMVPSU",
        "SDMVSTRA"
    ],
    
    "GHB": [
        "SEQN",
        "LBXGH"
    ],
    
    "BMX": [
        "SEQN",
        "BMXBMI",
        "BMXWAIST",
        "BMXWT"
    ],
    
    "DIQ": [
        "SEQN",
        "DIQ010"
    ],
    
    "PAQ": [
        "SEQN",
        "PAQ650",
        "PAQ655",
        "PAD660",
        "PAQ665",
        "PAQ670",
        "PAD675",
        "PAD680"
    ],
    
    "SMQ": [
        "SEQN",
        "SMQ020",
        "SMQ040"
    ],
    
    "BPQ": [
        "SEQN",
        "BPQ020",
        "BPQ080"
    ]
}

selected_variables

{'DEMO': ['SEQN',
  'RIDAGEYR',
  'RIAGENDR',
  'RIDRETH3',
  'DMDEDUC2',
  'INDFMPIR',
  'WTMEC2YR',
  'SDMVPSU',
  'SDMVSTRA'],
 'GHB': ['SEQN', 'LBXGH'],
 'BMX': ['SEQN', 'BMXBMI', 'BMXWAIST', 'BMXWT'],
 'DIQ': ['SEQN', 'DIQ010'],
 'PAQ': ['SEQN',
  'PAQ650',
  'PAQ655',
  'PAD660',
  'PAQ665',
  'PAQ670',
  'PAD675',
  'PAD680'],
 'SMQ': ['SEQN', 'SMQ020', 'SMQ040'],
 'BPQ': ['SEQN', 'BPQ020', 'BPQ080']}

In [132]:
# Sleep duration uses different variable names across cycles

selected_sleep_variables = {
    "2011-2012": ["SEQN", "SLD010H"],
    "2013-2014": ["SEQN", "SLD010H"],
    "2015-2016": ["SEQN", "SLD012"],
    "2017-2018": ["SEQN", "SLD012"]
}

selected_sleep_variables

{'2011-2012': ['SEQN', 'SLD010H'],
 '2013-2014': ['SEQN', 'SLD010H'],
 '2015-2016': ['SEQN', 'SLD012'],
 '2017-2018': ['SEQN', 'SLD012']}

In [134]:
# the variables needed from the 2011-2012 datasets

demo_g_selected = demo_g[selected_variables["DEMO"]].copy()
ghb_g_selected  = ghb_g[selected_variables["GHB"]].copy()
bmx_g_selected  = bmx_g[selected_variables["BMX"]].copy()
diq_g_selected  = diq_g[selected_variables["DIQ"]].copy()
paq_g_selected  = paq_g[selected_variables["PAQ"]].copy()
smq_g_selected  = smq_g[selected_variables["SMQ"]].copy()
bpq_g_selected  = bpq_g[selected_variables["BPQ"]].copy()

# Sleep is handled separately because the variable name changes by cycle
slq_g_selected = slq_g[
    selected_sleep_variables["2011-2012"]
].copy()

In [136]:
demo_h_selected = demo_h[selected_variables["DEMO"]].copy()
ghb_h_selected  = ghb_h[selected_variables["GHB"]].copy()
bmx_h_selected  = bmx_h[selected_variables["BMX"]].copy()
diq_h_selected  = diq_h[selected_variables["DIQ"]].copy()
paq_h_selected  = paq_h[selected_variables["PAQ"]].copy()
smq_h_selected  = smq_h[selected_variables["SMQ"]].copy()
bpq_h_selected  = bpq_h[selected_variables["BPQ"]].copy()

slq_h_selected = slq_h[
    selected_sleep_variables["2013-2014"]
].copy()

In [138]:
demo_i_selected = demo_i[selected_variables["DEMO"]].copy()
ghb_i_selected  = ghb_i[selected_variables["GHB"]].copy()
bmx_i_selected  = bmx_i[selected_variables["BMX"]].copy()
diq_i_selected  = diq_i[selected_variables["DIQ"]].copy()
paq_i_selected  = paq_i[selected_variables["PAQ"]].copy()
smq_i_selected  = smq_i[selected_variables["SMQ"]].copy()
bpq_i_selected  = bpq_i[selected_variables["BPQ"]].copy()

slq_i_selected = slq_i[
    selected_sleep_variables["2015-2016"]
].copy()

In [140]:
demo_j_selected = demo_j[selected_variables["DEMO"]].copy()
ghb_j_selected  = ghb_j[selected_variables["GHB"]].copy()
bmx_j_selected  = bmx_j[selected_variables["BMX"]].copy()
diq_j_selected  = diq_j[selected_variables["DIQ"]].copy()
paq_j_selected  = paq_j[selected_variables["PAQ"]].copy()
smq_j_selected  = smq_j[selected_variables["SMQ"]].copy()
bpq_j_selected  = bpq_j[selected_variables["BPQ"]].copy()

slq_j_selected = slq_j[
    selected_sleep_variables["2017-2018"]
].copy()

In [142]:
print("DEMO:", demo_g_selected.shape)
print("GHB :", ghb_g_selected.shape)
print("BMX :", bmx_g_selected.shape)
print("DIQ :", diq_g_selected.shape)
print("PAQ :", paq_g_selected.shape)
print("SMQ :", smq_g_selected.shape)
print("SLQ :", slq_g_selected.shape)
print("BPQ :", bpq_g_selected.shape)

DEMO: (9756, 9)
GHB : (6549, 2)
BMX : (9338, 4)
DIQ : (9364, 2)
PAQ : (9107, 8)
SMQ : (6790, 3)
SLQ : (6175, 2)
BPQ : (6175, 3)


# 5. Cross-Cycle Variable Standardization

The selected NHANES variables currently use the original CDC variable names. Before combining survey cycles, the variables will be renamed using more descriptive names.

In [144]:
# Standardized variable names

rename_map = {
    # Demographics
    "RIDAGEYR": "age",
    "RIAGENDR": "sex",
    "RIDRETH3": "race_ethnicity",
    "DMDEDUC2": "education",
    "INDFMPIR": "income_poverty_ratio",
    "WTMEC2YR": "mec_weight",
    "SDMVPSU": "psu",
    "SDMVSTRA": "stratum",

    # HbA1c
    "LBXGH": "hba1c",

    # Body measures
    "BMXBMI": "bmi",
    "BMXWAIST": "waist_cm",
    "BMXWT": "weight_kg",

    # Diabetes questionnaire
    "DIQ010": "diabetes_report",

    # Physical activity
    "PAQ650": "vigorous_activity",
    "PAQ655": "vigorous_days",
    "PAD660": "vigorous_minutes",
    "PAQ665": "moderate_activity",
    "PAQ670": "moderate_days",
    "PAD675": "moderate_minutes",
    "PAD680": "sedentary_minutes",

    # Smoking
    "SMQ020": "smoked_100_cigarettes",
    "SMQ040": "current_smoking",

    # Blood pressure / cholesterol
    "BPQ020": "hypertension",
    "BPQ080": "high_cholesterol"
}

# Rename variables for 2011-2012

demo_g_selected = demo_g_selected.rename(columns=rename_map)
ghb_g_selected = ghb_g_selected.rename(columns=rename_map)
bmx_g_selected = bmx_g_selected.rename(columns=rename_map)
diq_g_selected = diq_g_selected.rename(columns=rename_map)
paq_g_selected = paq_g_selected.rename(columns=rename_map)
smq_g_selected = smq_g_selected.rename(columns=rename_map)
bpq_g_selected = bpq_g_selected.rename(columns=rename_map)

# Sleep uses SLD010H in this cycle
slq_g_selected = slq_g_selected.rename(
    columns={"SLD010H": "sleep_hours"}
)

# Rename variables for 2013-2014

demo_h_selected = demo_h_selected.rename(columns=rename_map)
ghb_h_selected = ghb_h_selected.rename(columns=rename_map)
bmx_h_selected = bmx_h_selected.rename(columns=rename_map)
diq_h_selected = diq_h_selected.rename(columns=rename_map)
paq_h_selected = paq_h_selected.rename(columns=rename_map)
smq_h_selected = smq_h_selected.rename(columns=rename_map)
bpq_h_selected = bpq_h_selected.rename(columns=rename_map)

# Sleep uses SLD010H in this cycle
slq_h_selected = slq_h_selected.rename(
    columns={"SLD010H": "sleep_hours"}
)

# Rename variables for 2015-2016

demo_i_selected = demo_i_selected.rename(columns=rename_map)
ghb_i_selected = ghb_i_selected.rename(columns=rename_map)
bmx_i_selected = bmx_i_selected.rename(columns=rename_map)
diq_i_selected = diq_i_selected.rename(columns=rename_map)
paq_i_selected = paq_i_selected.rename(columns=rename_map)
smq_i_selected = smq_i_selected.rename(columns=rename_map)
bpq_i_selected = bpq_i_selected.rename(columns=rename_map)

# Sleep uses SLD012 in this cycle
slq_i_selected = slq_i_selected.rename(
    columns={"SLD012": "sleep_hours"}
)

# Rename variables for 2017-2018

demo_j_selected = demo_j_selected.rename(columns=rename_map)
ghb_j_selected = ghb_j_selected.rename(columns=rename_map)
bmx_j_selected = bmx_j_selected.rename(columns=rename_map)
diq_j_selected = diq_j_selected.rename(columns=rename_map)
paq_j_selected = paq_j_selected.rename(columns=rename_map)
smq_j_selected = smq_j_selected.rename(columns=rename_map)
bpq_j_selected = bpq_j_selected.rename(columns=rename_map)

# Sleep uses SLD012 in this cycle
slq_j_selected = slq_j_selected.rename(
    columns={"SLD012": "sleep_hours"}
)

In [146]:
print("2011-2012:", slq_g_selected.columns.tolist())
print("2013-2014:", slq_h_selected.columns.tolist())
print("2015-2016:", slq_i_selected.columns.tolist())
print("2017-2018:", slq_j_selected.columns.tolist())

2011-2012: ['SEQN', 'sleep_hours']
2013-2014: ['SEQN', 'sleep_hours']
2015-2016: ['SEQN', 'sleep_hours']
2017-2018: ['SEQN', 'sleep_hours']


In [148]:
# Add survey-cycle identifier

demo_g_selected["cycle"] = "2011-2012"
demo_h_selected["cycle"] = "2013-2014"
demo_i_selected["cycle"] = "2015-2016"
demo_j_selected["cycle"] = "2017-2018"

In [150]:
demo_g_selected.head(5)

,SEQN,age,sex,race_ethnicity,education,income_poverty_ratio,mec_weight,psu,stratum,cycle
0,62161.0,22.0,1.0,3.0,3.0,3.15,104236.582554,1.0,91.0,2011-2012
1,62162.0,3.0,2.0,1.0,NaN,0.60,16116.354010,3.0,92.0,2011-2012
2,62163.0,14.0,1.0,6.0,NaN,4.07,7869.485117,3.0,90.0,2011-2012
3,62164.0,44.0,2.0,3.0,4.0,1.67,127965.226204,1.0,94.0,2011-2012
4,62165.0,14.0,2.0,4.0,NaN,0.57,13384.042162,2.0,90.0,2011-2012


In [152]:
print("DEMO:")
print(demo_g_selected.columns.tolist())

print("\nGHB:")
print(ghb_g_selected.columns.tolist())

print("\nBMX:")
print(bmx_g_selected.columns.tolist())

print("\nDIQ:")
print(diq_g_selected.columns.tolist())

print("\nPAQ:")
print(paq_g_selected.columns.tolist())

print("\nSMQ:")
print(smq_g_selected.columns.tolist())

print("\nSLQ:")
print(slq_g_selected.columns.tolist())

print("\nBPQ:")
print(bpq_g_selected.columns.tolist())

DEMO:
['SEQN', 'age', 'sex', 'race_ethnicity', 'education', 'income_poverty_ratio', 'mec_weight', 'psu', 'stratum', 'cycle']

GHB:
['SEQN', 'hba1c']

BMX:
['SEQN', 'bmi', 'waist_cm', 'weight_kg']

DIQ:
['SEQN', 'diabetes_report']

PAQ:
['SEQN', 'vigorous_activity', 'vigorous_days', 'vigorous_minutes', 'moderate_activity', 'moderate_days', 'moderate_minutes', 'sedentary_minutes']

SMQ:
['SEQN', 'smoked_100_cigarettes', 'current_smoking']

SLQ:
['SEQN', 'sleep_hours']

BPQ:
['SEQN', 'hypertension', 'high_cholesterol']


###  Cleaning Special Response Codes

In [154]:
# Special NHANES response codes for the selected variables
# These codes represent responses such as Refused or Don't know.

special_code_map = {

    # Demographics
    "education": [7, 9],

    # Diabetes questionnaire
    "diabetes_report": [7, 9],

    # Physical activity
    "vigorous_activity": [7, 9],
    "vigorous_days": [77, 99],
    "vigorous_minutes": [7777, 9999],

    "moderate_activity": [7, 9],
    "moderate_days": [77, 99],
    "moderate_minutes": [7777, 9999],

    "sedentary_minutes": [7777, 9999],

    # Smoking
    "smoked_100_cigarettes": [7, 9],
    "current_smoking": [7, 9],

    # Sleep
    "sleep_hours": [77, 99],

    # Blood pressure / cholesterol
    "hypertension": [7, 9],
    "high_cholesterol": [7, 9]
}



In [156]:
def clean_special_codes(df):
    """
    Replace NHANES special response codes with NaN
    for the selected analysis variables.
    """

    df = df.copy()

    # Replace variable-specific special codes
    for column, codes in special_code_map.items():

        if column in df.columns:
            df[column] = df[column].replace(codes, np.nan)

    # PAD680 can contain a tiny floating-point number
    # after importing the SAS XPT file.
    # Values extremely close to zero are treated as 0.
    if "sedentary_minutes" in df.columns:

        df.loc[
            df["sedentary_minutes"].abs() < 1e-10,
            "sedentary_minutes"
        ] = 0

    return df

In [158]:
def clean_four_cycles(g, h, i, j):
    """
    Apply special-code cleaning to the same NHANES
    component across all four survey cycles.
    """

    return (
        clean_special_codes(g),
        clean_special_codes(h),
        clean_special_codes(i),
        clean_special_codes(j)
    )

In [160]:
# DEMO
demo_g_selected, demo_h_selected, demo_i_selected, demo_j_selected = (
    clean_four_cycles(
        demo_g_selected,
        demo_h_selected,
        demo_i_selected,
        demo_j_selected
    )
)

# DIQ
diq_g_selected, diq_h_selected, diq_i_selected, diq_j_selected = (
    clean_four_cycles(
        diq_g_selected,
        diq_h_selected,
        diq_i_selected,
        diq_j_selected
    )
)

# PAQ
paq_g_selected, paq_h_selected, paq_i_selected, paq_j_selected = (
    clean_four_cycles(
        paq_g_selected,
        paq_h_selected,
        paq_i_selected,
        paq_j_selected
    )
)

# SMQ
smq_g_selected, smq_h_selected, smq_i_selected, smq_j_selected = (
    clean_four_cycles(
        smq_g_selected,
        smq_h_selected,
        smq_i_selected,
        smq_j_selected
    )
)

# SLQ
slq_g_selected, slq_h_selected, slq_i_selected, slq_j_selected = (
    clean_four_cycles(
        slq_g_selected,
        slq_h_selected,
        slq_i_selected,
        slq_j_selected
    )
)

# BPQ
bpq_g_selected, bpq_h_selected, bpq_i_selected, bpq_j_selected = (
    clean_four_cycles(
        bpq_g_selected,
        bpq_h_selected,
        bpq_i_selected,
        bpq_j_selected
    )
)

In [ ]:
#  all  datasets together temporarily for validation

all_selected_datasets = [
    demo_g_selected, demo_h_selected,
    demo_i_selected, demo_j_selected,

    diq_g_selected, diq_h_selected,
    diq_i_selected, diq_j_selected,

    paq_g_selected, paq_h_selected,
    paq_i_selected, paq_j_selected,

    smq_g_selected, smq_h_selected,
    smq_i_selected, smq_j_selected,

    slq_g_selected, slq_h_selected,
    slq_i_selected, slq_j_selected,

    bpq_g_selected, bpq_h_selected,
    bpq_i_selected, bpq_j_selected
]


# see if any special codes remain

for variable, codes in special_code_map.items():

    remaining = 0

    for df in all_selected_datasets:

        if variable in df.columns:
            remaining += df[variable].isin(codes).sum()

    print(f"{variable:<25} remaining special codes: {remaining}")

###  Creating Physical-Activity Features

The NHANES Physical Activity Questionnaire records recreational activity using
multiple related variables.

For vigorous recreational activity:

- vigorous_activity indicates whether the participant performs vigorous
  recreational activity.
- vigorous_days records the number of days per week.
- vigorous_minutes records the number of minutes performed on a typical day.

For moderate recreational activity:

- moderate_activity indicates whether the participant performs moderate
  recreational activity.
- moderate_days records the number of days per week.
- moderate_minutes records the number of minutes performed on a typical day.

To make these variables easier to use in exploratory analysis and machine learning, weekly activity measures will be calculated as:

`days per week × minutes per day`

Participants who report that they do not perform the activity will receive
0 minutes per week.

If a participant reports performing the activity but the number of days or minutes is missing, the derived weekly value will remain missing rather than being assumed to be zero.

sedentary_minutes will remain as minutes per day.

In [162]:
def create_activity_features(df):
    """
    Create weekly vigorous and moderate recreational
    physical-activity features.
    """

    df = df.copy()

    # Vigorous recreational activity

    #  missing values
    df["vigorous_minutes_week"] = np.nan

    # PAQ650 = 2 means participant reported NO vigorous activity
    df.loc[
        df["vigorous_activity"] == 2,
        "vigorous_minutes_week"
    ] = 0

    # PAQ650 = 1 means YES.
    # Calculating weekly minutes only when both days and minutes are available.
    vigorous_yes = (
        (df["vigorous_activity"] == 1)
        & df["vigorous_days"].notna()
        & df["vigorous_minutes"].notna()
    )

    df.loc[
        vigorous_yes,
        "vigorous_minutes_week"
    ] = (
        df.loc[vigorous_yes, "vigorous_days"]
        * df.loc[vigorous_yes, "vigorous_minutes"]
    )

    # Moderate recreational activity

    df["moderate_minutes_week"] = np.nan

    # PAQ665 = 2 means participant reported NO moderate activity
    df.loc[
        df["moderate_activity"] == 2,
        "moderate_minutes_week"
    ] = 0

    # Calculating weekly minutes for participants reporting YES
    moderate_yes = (
        (df["moderate_activity"] == 1)
        & df["moderate_days"].notna()
        & df["moderate_minutes"].notna()
    )

    df.loc[
        moderate_yes,
        "moderate_minutes_week"
    ] = (
        df.loc[moderate_yes, "moderate_days"]
        * df.loc[moderate_yes, "moderate_minutes"]
    )

    return df

In [164]:
paq_g_selected = create_activity_features(paq_g_selected)
paq_h_selected = create_activity_features(paq_h_selected)
paq_i_selected = create_activity_features(paq_i_selected)
paq_j_selected = create_activity_features(paq_j_selected)

In [166]:
paq_g_selected[
    [
        "vigorous_activity",
        "vigorous_days",
        "vigorous_minutes",
        "vigorous_minutes_week",
        "moderate_activity",
        "moderate_days",
        "moderate_minutes",
        "moderate_minutes_week",
        "sedentary_minutes"
    ]
].head(10)

,vigorous_activity,vigorous_days,vigorous_minutes,vigorous_minutes_week,moderate_activity,moderate_days,moderate_minutes,moderate_minutes_week,sedentary_minutes
0,2.0,NaN,NaN,0.0,2.0,NaN,NaN,0.0,300.0
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2.0,NaN,NaN,0.0,2.0,NaN,NaN,0.0,720.0
3,1.0,5.0,60.0,300.0,1.0,1.0,45.0,45.0,300.0
4,1.0,2.0,25.0,50.0,1.0,2.0,25.0,50.0,600.0
5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,2.0,NaN,NaN,0.0,2.0,NaN,NaN,0.0,360.0
8,1.0,5.0,90.0,450.0,2.0,NaN,NaN,0.0,600.0
9,1.0,5.0,20.0,100.0,1.0,4.0,30.0,120.0,480.0


In [168]:
activity_features = [
    "vigorous_minutes_week",
    "moderate_minutes_week",
    "sedentary_minutes"
]

paq_g_selected[activity_features].describe()

,vigorous_minutes_week,moderate_minutes_week,sedentary_minutes
count,6774.000000,6773.000000,6742.000000
mean,81.974461,98.444707,383.785820
std,207.388926,204.829967,203.043619
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,240.000000
50%,0.000000,0.000000,360.000000
75%,60.000000,120.000000,480.000000
max,4200.000000,3360.000000,1380.000000


In [170]:
paq_datasets_selected = {
    "2011-2012": paq_g_selected,
    "2013-2014": paq_h_selected,
    "2015-2016": paq_i_selected,
    "2017-2018": paq_j_selected
}

for cycle, df in paq_datasets_selected.items():

    print(f"\n{cycle}")
    print("-" * 40)

    print(
        df[
            [
                "vigorous_minutes_week",
                "moderate_minutes_week",
                "sedentary_minutes"
            ]
        ].describe()
    )


2011-2012
----------------------------------------
       vigorous_minutes_week  moderate_minutes_week  sedentary_minutes
count            6774.000000            6773.000000        6742.000000
mean               81.974461              98.444707         383.785820
std               207.388926             204.829967         203.043619
min                 0.000000               0.000000           0.000000
25%                 0.000000               0.000000         240.000000
50%                 0.000000               0.000000         360.000000
75%                60.000000             120.000000         480.000000
max              4200.000000            3360.000000        1380.000000

2013-2014
----------------------------------------
       vigorous_minutes_week  moderate_minutes_week  sedentary_minutes
count            7139.000000            7140.000000        7109.000000
mean               81.116963              93.773529         438.685188
std               193.412730             206

In [175]:
for df in [
    paq_g_selected,
    paq_h_selected,
    paq_i_selected,
    paq_j_selected
]:

    df["recreational_activity_minutes_week"] = (
        df["vigorous_minutes_week"]
        + df["moderate_minutes_week"]
    )

#### Physical-Activity Feature Findings

Weekly vigorous and moderate recreational activity measures were successfully created from the NHANES physical-activity questionnaire variables.

Participants who reported no vigorous or moderate recreational activity were assigned 0 minutes per week. For participants reporting activity, weekly minutes were calculated by multiplying the number of active days per week by the reported minutes of activity per day.

The distributions show that many participants reported no recreational vigorous or moderate activity, as indicated by median vigorous activity values of 0 minutes per week across all four survey cycles.

Sedentary time was also successfully cleaned and retained as minutes per day. The median sedentary time ranged from approximately 300 to 480 minutes per day across the four cycles.

Some very high weekly activity values are present. These observations will be examined during exploratory data analysis rather than automatically removed, because they may represent valid but extreme responses.

The derived physical-activity measures can now be used as behavioral predictors in later analysis.

In [177]:
#  total recreational activity minutes per week
# The result remains missing if either component is missing.

for df in [
    paq_g_selected,
    paq_h_selected,
    paq_i_selected,
    paq_j_selected
]:
    df["recreational_activity_minutes_week"] = (
        df["vigorous_minutes_week"]
        + df["moderate_minutes_week"]
    )

In [179]:
for cycle, df in paq_datasets_selected.items():

    print(f"\n{cycle}")

    print(
        df[
            [
                "vigorous_minutes_week",
                "moderate_minutes_week",
                "recreational_activity_minutes_week",
                "sedentary_minutes"
            ]
        ].describe()
    )


2011-2012
       vigorous_minutes_week  moderate_minutes_week  \
count            6774.000000            6773.000000   
mean               81.974461              98.444707   
std               207.388926             204.829967   
min                 0.000000               0.000000   
25%                 0.000000               0.000000   
50%                 0.000000               0.000000   
75%                60.000000             120.000000   
max              4200.000000            3360.000000   

       recreational_activity_minutes_week  sedentary_minutes  
count                         6772.000000        6742.000000  
mean                           180.452008         383.785820  
std                            310.178380         203.043619  
min                              0.000000           0.000000  
25%                              0.000000         240.000000  
50%                             50.000000         360.000000  
75%                            240.000000         48

## 6. Combining Survey Cycles and Merging Components


After selecting, cleaning, and standardizing the variables, the four NHANES survey cycles can be combined.

First, the same component from each survey cycle will be stacked vertically. For example, the four demographic datasets will be combined into one demographic dataset.

After each component has been combined across cycles, the components will be merged using `SEQN`, the unique NHANES participant identifier.

The demographic dataset will serve as the base dataset so that participants are retained even when information from another component is missing.

In [181]:
# Combine the four survey cycles within each component

demo_all = pd.concat(
    [
        demo_g_selected,
        demo_h_selected,
        demo_i_selected,
        demo_j_selected
    ],
    ignore_index=True
)

ghb_all = pd.concat(
    [
        ghb_g_selected,
        ghb_h_selected,
        ghb_i_selected,
        ghb_j_selected
    ],
    ignore_index=True
)

bmx_all = pd.concat(
    [
        bmx_g_selected,
        bmx_h_selected,
        bmx_i_selected,
        bmx_j_selected
    ],
    ignore_index=True
)

diq_all = pd.concat(
    [
        diq_g_selected,
        diq_h_selected,
        diq_i_selected,
        diq_j_selected
    ],
    ignore_index=True
)

paq_all = pd.concat(
    [
        paq_g_selected,
        paq_h_selected,
        paq_i_selected,
        paq_j_selected
    ],
    ignore_index=True
)

smq_all = pd.concat(
    [
        smq_g_selected,
        smq_h_selected,
        smq_i_selected,
        smq_j_selected
    ],
    ignore_index=True
)

slq_all = pd.concat(
    [
        slq_g_selected,
        slq_h_selected,
        slq_i_selected,
        slq_j_selected
    ],
    ignore_index=True
)

bpq_all = pd.concat(
    [
        bpq_g_selected,
        bpq_h_selected,
        bpq_i_selected,
        bpq_j_selected
    ],
    ignore_index=True
)

print("DEMO:", demo_all.shape)
print("GHB :", ghb_all.shape)
print("BMX :", bmx_all.shape)
print("DIQ :", diq_all.shape)
print("PAQ :", paq_all.shape)
print("SMQ :", smq_all.shape)
print("SLQ :", slq_all.shape)
print("BPQ :", bpq_all.shape)

DEMO: (39156, 10)
GHB : (26673, 2)
BMX : (37399, 4)
DIQ : (37606, 2)
PAQ : (33702, 11)
SMQ : (27683, 3)
SLQ : (25127, 2)
BPQ : (25127, 3)


In [183]:
combined_components = {
    "DEMO": demo_all,
    "GHB": ghb_all,
    "BMX": bmx_all,
    "DIQ": diq_all,
    "PAQ": paq_all,
    "SMQ": smq_all,
    "SLQ": slq_all,
    "BPQ": bpq_all
}

for name, df in combined_components.items():

    duplicates = df["SEQN"].duplicated().sum()

    print(
        f"{name:<5} rows: {len(df):>6} "
        f"duplicate SEQN: {duplicates}"
    )

DEMO  rows:  39156 duplicate SEQN: 0
GHB   rows:  26673 duplicate SEQN: 0
BMX   rows:  37399 duplicate SEQN: 0
DIQ   rows:  37606 duplicate SEQN: 0
PAQ   rows:  33702 duplicate SEQN: 0
SMQ   rows:  27683 duplicate SEQN: 0
SLQ   rows:  25127 duplicate SEQN: 0
BPQ   rows:  25127 duplicate SEQN: 0


In [185]:
#  demographics as the base participant dataset

nhanes = demo_all.copy()

# Merge laboratory HbA1c
nhanes = nhanes.merge(
    ghb_all,
    on="SEQN",
    how="left"
)

# Merge body measurements
nhanes = nhanes.merge(
    bmx_all,
    on="SEQN",
    how="left"
)

# Merge diabetes questionnaire
nhanes = nhanes.merge(
    diq_all,
    on="SEQN",
    how="left"
)

# Merge physical activity
nhanes = nhanes.merge(
    paq_all,
    on="SEQN",
    how="left"
)

# Merge smoking
nhanes = nhanes.merge(
    smq_all,
    on="SEQN",
    how="left"
)

# Merge sleep
nhanes = nhanes.merge(
    slq_all,
    on="SEQN",
    how="left"
)

# Merge blood pressure and cholesterol
nhanes = nhanes.merge(
    bpq_all,
    on="SEQN",
    how="left"
)

print("Final merged shape:", nhanes.shape)

Final merged shape: (39156, 30)


In [187]:
print("Rows:", nhanes.shape[0])
print("Columns:", nhanes.shape[1])

print(
    "Unique participants:",
    nhanes["SEQN"].nunique()
)

print(
    "Duplicate participants:",
    nhanes["SEQN"].duplicated().sum()
)

Rows: 39156
Columns: 30
Unique participants: 39156
Duplicate participants: 0


In [189]:
nhanes.head()

,SEQN,age,sex,race_ethnicity,education,income_poverty_ratio,mec_weight,psu,stratum,cycle,...,moderate_minutes,sedentary_minutes,vigorous_minutes_week,moderate_minutes_week,recreational_activity_minutes_week,smoked_100_cigarettes,current_smoking,sleep_hours,hypertension,high_cholesterol
0,62161.0,22.0,1.0,3.0,3.0,3.15,104236.582554,1.0,91.0,2011-2012,...,NaN,300.0,0.0,0.0,0.0,2.0,NaN,8.0,2.0,2.0
1,62162.0,3.0,2.0,1.0,NaN,0.60,16116.354010,3.0,92.0,2011-2012,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,62163.0,14.0,1.0,6.0,NaN,4.07,7869.485117,3.0,90.0,2011-2012,...,NaN,720.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN
3,62164.0,44.0,2.0,3.0,4.0,1.67,127965.226204,1.0,94.0,2011-2012,...,45.0,300.0,300.0,45.0,345.0,2.0,NaN,8.0,2.0,2.0
4,62165.0,14.0,2.0,4.0,NaN,0.57,13384.042162,2.0,90.0,2011-2012,...,25.0,600.0,50.0,50.0,100.0,NaN,NaN,NaN,NaN,NaN


In [191]:
nhanes.columns.tolist()

['SEQN',
 'age',
 'sex',
 'race_ethnicity',
 'education',
 'income_poverty_ratio',
 'mec_weight',
 'psu',
 'stratum',
 'cycle',
 'hba1c',
 'bmi',
 'waist_cm',
 'weight_kg',
 'diabetes_report',
 'vigorous_activity',
 'vigorous_days',
 'vigorous_minutes',
 'moderate_activity',
 'moderate_days',
 'moderate_minutes',
 'sedentary_minutes',
 'vigorous_minutes_week',
 'moderate_minutes_week',
 'recreational_activity_minutes_week',
 'smoked_100_cigarettes',
 'current_smoking',
 'sleep_hours',
 'hypertension',
 'high_cholesterol']

### Combined Dataset Findings

The four NHANES survey cycles were successfully combined and the eight data
components were merged using `SEQN`, the unique participant identifier.

The resulting dataset contains 39,156 rows and 30 columns.

Each row represents one unique NHANES participant, and no duplicate `SEQN`
values were found after the merge.

The demographic dataset was used as the base dataset and left joins were used
to preserve participants even when information from another NHANES component
was unavailable.

The final merged dataset therefore follows a tidy structure:

- one row represents one participant;
- one column represents one participant characteristic;
- `SEQN` uniquely identifies each participant.

Differences in the number of observations across components are expected
because NHANES examinations and questionnaires have different eligibility
criteria and patterns of missing data.

## 7. Initial Exploration

Initial exploratory analysis will focus on the study population, target
variable, selected predictors, and missing data.

Because the project focuses on adults without previously diagnosed diabetes,
the first step is to examine how the sample changes after applying the planned
eligibility criteria.

### Defining the Analysis Population

The analysis will focus on adults who have a valid laboratory HbA1c measurement
and who do not report a previous diabetes diagnosis. 

For the primary analysis, participants must meet the following criteria:

1. Age 20 years or older.
2. A valid HbA1c measurement is available.
3. `DIQ010 = 2`, indicating that the participant reported not having been told
   by a doctor or health professional that they have diabetes.

Participants reporting diagnosed diabetes (`DIQ010 = 1`) will be excluded.

Participants reporting "borderline" diabetes (`DIQ010 = 3`) will also be
excluded from the primary analysis because they may already be aware of
abnormal glycemic status. They may be considered later in a sensitivity
analysis.

Missing predictor values will not be used as an exclusion criterion at this
stage. Missingness will be evaluated separately.

In [194]:
nhanes["diabetes_report"].value_counts(
    dropna=False
).sort_index()

diabetes_report
1.0     3194
2.0    33749
3.0      641
NaN     1572
Name: count, dtype: int64

In [196]:
# Total merged NHANES participants
n_total = len(nhanes)

# Adults age 20+
adults = nhanes[
    nhanes["age"] >= 20
].copy()

n_adults = len(adults)

# Adults with a valid HbA1c measurement
adults_hba1c = adults[
    adults["hba1c"].notna()
].copy()

n_adults_hba1c = len(adults_hba1c)

# Final primary analysis population:
# DIQ010 = participant reported NO diabetes
analysis_data = adults_hba1c[
    adults_hba1c["diabetes_report"] == 2
].copy()

n_analysis = len(analysis_data)

In [198]:
print("Sample Selection")
print("-" * 50)

print(f"Total merged participants:                 {n_total}")
print(f"Adults age 20+:                            {n_adults}")
print(f"Adults age 20+ with valid HbA1c:           {n_adults_hba1c}")
print(f"Final analysis population (DIQ010 = 2):    {n_analysis}")

Sample Selection
--------------------------------------------------
Total merged participants:                 39156
Adults age 20+:                            22617
Adults age 20+ with valid HbA1c:           20631
Final analysis population (DIQ010 = 2):    17156


In [204]:
diabetes_status_counts = (
    adults_hba1c["diabetes_report"]
    .value_counts(dropna=False)
    .sort_index()
)
diabetes_status_labels = {
    1.0: "Yes - Reported diabetes",
    2.0: "No reported diabetes",
    3.0: "Borderline"
}

status_summary = (
    adults_hba1c["diabetes_report"]
    .map(diabetes_status_labels)
    .fillna("Missing / Unknown")
    .value_counts()
)

status_summary

diabetes_report
No reported diabetes       17156
Yes - Reported diabetes     2913
Borderline                   548
Missing / Unknown             14
Name: count, dtype: int64

In [206]:
print("Final analysis dataset shape:", analysis_data.shape)

print(
    "Age range:",
    analysis_data["age"].min(),
    "to",
    analysis_data["age"].max()
)

print(
    "Missing HbA1c:",
    analysis_data["hba1c"].isna().sum()
)

print(
    "Diabetes-report values remaining:",
    analysis_data["diabetes_report"].unique()
)

Final analysis dataset shape: (17156, 30)
Age range: 20.0 to 80.0
Missing HbA1c: 0
Diabetes-report values remaining: [2.]


In [208]:
analysis_data["cycle"].value_counts().sort_index()
(
    analysis_data["cycle"]
    .value_counts(normalize=True)
    .sort_index()
    * 100
).round(2)

cycle
2011-2012    24.88
2013-2014    26.49
2015-2016    25.02
2017-2018    23.61
Name: proportion, dtype: float64

**Here, we see that 17,156 NHANES participants age 20 years or older, with a valid HbA1c measurement, who explicitly reported that they had not been told they have diabetes (DIQ010 = 2) among the 20,631 adults with valid HbA1c.**

Participants reporting diabetes (`DIQ010 = 1`), borderline diabetes
(`DIQ010 = 3`), or missing/unknown diabetes status were excluded

#### Defining the HbA1c target

HbA1c < 5.7%   → 0 = Normal

HbA1c ≥ 5.7%   → 1 = Elevated

HbA1c itself will not be included as a predictor. The resulting binary variable, `elevated_hba1c`, will serve as the target for machine-learning models.

In [222]:
analysis_data["elevated_hba1c"] = (
    analysis_data["hba1c"] >= 5.7
).astype(int)

analysis_data[
    ["hba1c", "elevated_hba1c"]
].head(10)

,hba1c,elevated_hba1c
0,5.1,0
3,4.9,0
8,5.4,0
11,5.6,0
13,5.0,0
15,5.4,0
16,5.8,1
17,5.7,1
18,5.7,1
19,5.3,0


In [224]:
target_counts = (
    analysis_data["elevated_hba1c"]
    .value_counts()
    .sort_index()
)

target_percent = (
    analysis_data["elevated_hba1c"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)

target_summary = pd.DataFrame({
    "count": target_counts,
    "percent": target_percent
})

target_summary.index = [
    "Normal HbA1c",
    "Elevated HbA1c"
]

target_summary

,count,percent
Normal HbA1c,11693,68.16
Elevated HbA1c,5463,31.84


Among the 17,156 participants in the primary analysis population, 11,693
participants (68.16%) have normal HbA1c and 5,463 participants (31.84%) have
elevated HbA1c. 

###  Missing Data Analysis

In [228]:
predictor_columns = [
    "age",
    "sex",
    "race_ethnicity",
    "education",
    "income_poverty_ratio",
    "bmi",
    "waist_cm",
    "weight_kg",
    "vigorous_minutes_week",
    "moderate_minutes_week",
    "recreational_activity_minutes_week",
    "sedentary_minutes",
    "smoked_100_cigarettes",
    "current_smoking",
    "sleep_hours",
    "hypertension",
    "high_cholesterol"
]

missing_summary = pd.DataFrame({
    "missing_count": analysis_data[predictor_columns].isna().sum(),
    "missing_percent": (
        analysis_data[predictor_columns].isna().mean() * 100
    ).round(2)
})

missing_summary = missing_summary.sort_values(
    "missing_percent",
    ascending=False
)

missing_summary


,missing_count,missing_percent
current_smoking,10054,58.60
income_poverty_ratio,1629,9.50
waist_cm,816,4.76
bmi,193,1.12
weight_kg,166,0.97
sedentary_minutes,109,0.64
high_cholesterol,78,0.45
sleep_hours,60,0.35
recreational_activity_minutes_week,25,0.15
moderate_minutes_week,20,0.12


Most selected predictors have relatively low levels of missing data, currently smoking shows the highest observation of missing data which might be because it's a follow up question to if they have smoked at least 100 cigerates in lifetime. Other variables such as income income_poverty_ratio, waist_cm, bmi are above 1% catogery and will be looked into later before final model training.

**Smoking status Feature**


In [237]:
analysis_data["smoking_status"] = pd.Series(
    pd.NA,
    index=analysis_data.index,
    dtype="string"
)

# Never smoker:
analysis_data.loc[
    analysis_data["smoked_100_cigarettes"] == 2,
    "smoking_status"
] = "Never"

# Current smoker:
analysis_data.loc[
    (analysis_data["smoked_100_cigarettes"] == 1) &
    (analysis_data["current_smoking"].isin([1, 2])),
    "smoking_status"
] = "Current"

# Former smoker:
analysis_data.loc[
    (analysis_data["smoked_100_cigarettes"] == 1) &
    (analysis_data["current_smoking"] == 3),
    "smoking_status"
] = "Former"
analysis_data["smoking_status"].value_counts(
    dropna=False
)
print(
    "Missing smoking status:",
    analysis_data["smoking_status"].isna().sum()
)

Missing smoking status: 14


In [239]:
analysis_data["smoking_status"].value_counts(
    dropna=False
)

smoking_status
Never      10040
Former      3656
Current     3446
<NA>          14
Name: count, dtype: Int64

**Among the 17,156 participants in the analysis population:**

- 10,040 were classified as `Never` smokers
- 3,656 were classified as `Former` smokers
- 3,446 were classified as `Current` smokers
- 14 participants had missing smoking status

### Continuous Variable Distributions and Outlier Checks

The continuous predictors will be examined using summary statistics,
histograms, and boxplots. This will be helpful to identify unusual ranges, skewed distributions, and
potential extreme values before machine-learning preprocessing.

In [248]:
continuous_features = [
    "age",
    "income_poverty_ratio",
    "bmi",
    "waist_cm",
    "weight_kg",
    "vigorous_minutes_week",
    "moderate_minutes_week",
    "recreational_activity_minutes_week",
    "sedentary_minutes",
    "sleep_hours"
]
continuous_summary = (
    analysis_data[continuous_features]
    .describe()
    .T
)

continuous_summary["missing_count"] = (
    analysis_data[continuous_features]
    .isna()
    .sum()
)

continuous_summary["missing_percent"] = (
    analysis_data[continuous_features]
    .isna()
    .mean()
    .mul(100)
    .round(2)
)

continuous_summary

,count,mean,std,min,25%,50%,75%,max,missing_count,missing_percent
age,17156.0,47.416764,17.486420,2.000000e+01,32.00,46.00,61.000,80.0,0,0.00
income_poverty_ratio,15527.0,2.521682,1.642887,5.397605e-79,1.09,2.11,4.120,5.0,1629,9.50
bmi,16963.0,28.774150,6.876530,1.340000e+01,24.00,27.70,32.200,86.2,193,1.12
waist_cm,16340.0,97.942864,16.086828,5.550000e+01,86.40,96.60,107.400,177.9,816,4.76
weight_kg,16990.0,80.315150,21.272429,2.910000e+01,65.30,77.20,91.475,242.6,166,0.97
vigorous_minutes_week,17144.0,63.081778,161.867353,0.000000e+00,0.00,0.00,30.000,3360.0,12,0.07
moderate_minutes_week,17136.0,92.686041,197.952353,0.000000e+00,0.00,0.00,120.000,3360.0,20,0.12
recreational_activity_minutes_week,17131.0,155.794291,281.059721,0.000000e+00,0.00,20.00,210.000,3540.0,25,0.15
sedentary_minutes,17047.0,368.575878,200.415657,0.000000e+00,240.00,360.00,480.000,1320.0,109,0.64
sleep_hours,17096.0,7.236985,1.530658,2.000000e+00,6.00,7.00,8.000,14.5,60,0.35
